// @progress/kendo-vue-dateinputs/calendar/utils/main.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2025 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
import { isInRange as c } from "../../utils.mjs";
import { EMPTY_SELECTIONRANGE as s } from "../models/SelectionRange.mjs";
import { getDate as e } from "@progress/kendo-date-math";
import { isObject as a } from "@progress/kendo-vue-common";
const A = (t, n, r, o) => o !== void 0 ? o !== null && c(o, t, n) ? o : null : r !== null && c(r, t, n) ? r : null, f = (t, n, r) => r instanceof Date && !Array.isArray(r) && c(e(r), t, n) ? e(r) : null, y = (t, n, r) => Array.isArray(r) ? r.filter((o) => c(o, t, n)).map((o) => e(o)) : null, x = (t) => a(t) && !(t instanceof Date) && t !== null && !Array.isArray(t) ? t : s, D = (t, n, r) => t || n && n[0] || r && r.start, d = (t, n) => t.start === null && n === null ? "start" : t.end === null ? "end" : "start";
export {
  A as calculateValue,
  d as extractActiveRange,
  f as extractDateFromValue,
  D as extractFocusedDate,
  y as extractMultipleFromValue,
  x as extractRangeFromValue
};
