// @progress/kendo-vue-grid/drag/ColumnDraggable.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2025 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
import { defineComponent as a, createVNode as r } from "vue";
import { getDefaultSlots as n, Draggable as s } from "@progress/kendo-vue-common";
const i = /* @__PURE__ */ a({
  props: {
    onPressHandler: Function,
    onDragHandler: Function,
    onReleaseHandler: Function
  },
  mounted() {
    this.$el && (this.draggable = this.$refs.draggable);
  },
  methods: {
    onPress(t) {
      const e = this.draggable && this.draggable.element;
      e && this.$emit("pressHandler", t, e);
    },
    onDrag(t) {
      const e = this.draggable && this.draggable.element;
      e && this.$emit("dragHandler", t, e);
    },
    onRelease(t) {
      const e = this.draggable && this.draggable.element;
      e && this.$emit("releaseHandler", t, e);
    }
  },
  render() {
    const t = n(this);
    return r(s, {
      onPress: this.onPress,
      onDrag: this.onDrag,
      onRelease: this.onRelease,
      ref: "draggable"
    }, {
      default: () => [r("tr", {
        class: "k-table-row",
        role: "row"
      }, [t])]
    });
  }
});
export {
  i as ColumnDraggable
};
