// @progress/kendo-vue-layout/breadcrumb/BreadcrumbDelimiter.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2025 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
import { defineComponent as n, createVNode as o } from "vue";
import { Icon as t } from "@progress/kendo-vue-common";
import { chevronLeftIcon as d, chevronRightIcon as m } from "@progress/kendo-svg-icons";
const p = /* @__PURE__ */ n({
  name: "KendoBreadcrumbDelimiter",
  props: {
    id: String,
    dir: String,
    tabIndex: Number
  },
  render() {
    const {
      id: r,
      dir: e,
      tabIndex: i
    } = this.$props;
    return o(t, {
      id: r,
      "aria-hidden": !0,
      tabIndex: i || 0,
      size: "small",
      icon: e === "rtl" ? d : m,
      class: "k-breadcrumb-delimiter-icon"
    }, null);
  }
});
export {
  p as BreadcrumbDelimiter
};
