// @progress/kendo-vue-layout/breadcrumb/BreadcrumbListItem.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2025 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
import { defineComponent as r, createVNode as s } from "vue";
import { getDefaultSlots as i } from "@progress/kendo-vue-common";
const a = /* @__PURE__ */ r({
  name: "KendoBreadcrumbListItem",
  props: {
    id: String,
    isFirstItem: Boolean,
    isLastItem: Boolean
  },
  computed: {
    itemClasses() {
      const {
        isFirstItem: t,
        isLastItem: e
      } = this.$props;
      return {
        "k-breadcrumb-item": !0,
        "k-breadcrumb-root-item": t,
        "k-breadcrumb-last-item": e
      };
    }
  },
  render() {
    const t = i(this);
    return s("li", {
      id: this.$props.id,
      class: this.itemClasses
    }, [t]);
  }
});
export {
  a as BreadcrumbListItem
};
