// @progress/kendo-vue-layout/breadcrumb/BreadcrumbOrderedList.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2025 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
import { defineComponent as i, createVNode as n } from "vue";
import { getDefaultSlots as d, getTabIndex as s, getDir as a } from "@progress/kendo-vue-common";
const p = /* @__PURE__ */ i({
  name: "KendoBreadcrumbOrderedList",
  props: {
    rootItem: Boolean,
    id: String,
    dir: String,
    tabIndex: Number,
    disabled: Boolean,
    collapseMode: {
      type: String,
      validator: function(e) {
        return ["auto", "wrap", "none"].includes(e);
      }
    }
  },
  computed: {
    itemClasses() {
      const {
        rootItem: e,
        disabled: t,
        collapseMode: r
      } = this.$props;
      return {
        "k-breadcrumb-root-item-container": e,
        "k-breadcrumb-container": !e,
        "k-disabled": t,
        "!k-flex-wrap": r === "wrap",
        "k-flex-none": r === "none"
      };
    }
  },
  created() {
    this.currentDir = void 0;
  },
  mounted() {
    this.currentDir = a(this.$el, this.$props.dir);
  },
  render() {
    const e = d(this), {
      id: t,
      disabled: r,
      tabIndex: o
    } = this.$props;
    return n("ol", {
      id: t,
      dir: this.currentDir,
      tabindex: s(o, r),
      class: this.itemClasses
    }, [e]);
  }
});
export {
  p as BreadcrumbOrderedList
};
