// @progress/kendo-vue-inputs/signature/SignatureBottomActions.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2025 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
import { defineComponent as n, createVNode as e } from "vue";
import { Button as r } from "@progress/kendo-vue-buttons";
import { xIcon as s } from "@progress/kendo-svg-icons";
const m = /* @__PURE__ */ n({
  name: "KendoSignatureBottomActions",
  emits: {
    clear: null
  },
  props: {
    showClear: Boolean,
    size: String,
    title: String
  },
  methods: {
    onClear(t) {
      this.$emit("clear", t);
    }
  },
  render() {
    const {
      showClear: t,
      title: o,
      size: i
    } = this.$props;
    return e("div", {
      class: "k-signature-actions k-signature-actions-bottom"
    }, [t && e(r, {
      class: "k-signature-action k-signature-clear",
      icon: "x",
      svgIcon: s,
      shape: null,
      fillMode: "flat",
      size: i,
      onClick: this.onClear,
      "aria-label": o,
      title: o
    }, null)]);
  }
});
export {
  m as SignatureBottomActions
};
