// @progress/kendo-vue-dropdowns/MultiSelect/TagList.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2025 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
import { defineComponent as h, createVNode as i, h as v } from "vue";
import { getDefaultSlots as S, classNames as T, getTemplate as $, kendoThemeMaps as b } from "@progress/kendo-vue-common";
import { Chip as k } from "@progress/kendo-vue-buttons";
const x = (e) => e.preventDefault(), y = (e) => e.event.stopPropagation(), C = /* @__PURE__ */ h({
  name: "TagList",
  inheritAttrs: !1,
  props: {
    removeTagIcon: String,
    dataItems: Array,
    guid: String,
    focused: Object,
    tagRender: [String, Function, Object],
    tagsRounded: {
      type: String,
      default: "medium",
      validator: function(e) {
        return ["small", "medium", "large", "full"].includes(e);
      }
    },
    fillMode: {
      type: String,
      default: "solid",
      validator: function(e) {
        return ["solid", "flat", "outline"].includes(e);
      }
    },
    id: String,
    className: String,
    size: {
      type: String,
      default: "medium",
      validator: function(e) {
        return ["small", "medium", "large"].includes(e);
      }
    },
    onTagdelete: Function
  },
  computed: {
    wrapperClass() {
      const {
        size: e,
        disabled: t
      } = this.$props;
      return {
        "k-chip-list": !0,
        [`k-chip-list-${b.sizeMap[e] || e}`]: e,
        "k-disabled": t
      };
    }
  },
  methods: {
    onTagDelete(e, t) {
      this.$emit("tagdelete", e, t);
    }
  },
  render() {
    const e = S(this), {
      dataItems: t,
      guid: n,
      size: r,
      disabled: l,
      tagsRounded: d,
      removeTagIcon: u,
      id: p,
      className: c
    } = this.$props;
    return i("div", {
      class: T("k-input-values", c),
      id: p,
      style: this.$attrs.style
    }, [i("div", {
      class: this.wrapperClass,
      role: "listbox",
      id: "tagslist-" + n
    }, [t.map(function(s, o) {
      const m = `tag-${n}-${s.text.toString().replace(/\s+/g, "-")}`, a = s === this.$props.focused, g = i(k, {
        id: m,
        tabIndex: -1,
        "aria-selected": a,
        role: "option",
        "aria-setsize": t.length,
        key: s.text + o,
        text: s.text,
        disabled: l,
        removable: !0,
        onRemove: (f) => this.onTagDelete(s.data, f),
        removeIcon: u,
        onMousedown: x,
        onClick: y,
        class: a ? "k-focus" : void 0,
        rounded: d,
        size: r
      }, null);
      return $.call(this, {
        h: v,
        template: this.$props.tagRender,
        defaultRendering: g,
        additionalProps: {
          ...this.$props,
          tagData: s,
          index: o
        },
        additionalListeners: {
          tagdelete: this.onTagDelete
        }
      });
    }, this)]), e]);
  }
});
export {
  C as TagList
};
