// @progress/kendo-vue-labels/Error.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2025 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
import { defineComponent as t, createVNode as e } from "vue";
import { packageMetadata as o } from "./package-metadata.mjs";
import { getDefaultSlots as a, validatePackage as i } from "@progress/kendo-vue-common";
const p = /* @__PURE__ */ t({
  name: "KendoError",
  props: {
    id: String,
    direction: {
      type: String,
      default: "start",
      validator: function(r) {
        return ["start", "end"].includes(r);
      }
    }
  },
  created() {
    i(o);
  },
  computed: {
    errorClassName() {
      return {
        "k-form-error": !0,
        "k-text-start": this.$props.direction === "start",
        "k-text-end": this.$props.direction === "end"
      };
    }
  },
  render() {
    const r = a(this);
    return e("div", {
      id: this.$props.id,
      role: "alert",
      class: this.errorClassName
    }, [r]);
  }
});
export {
  p as Error
};
