// @progress/kendo-vue-labels/Hint.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2025 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
import { defineComponent as e, createVNode as r } from "vue";
import { packageMetadata as i } from "./package-metadata.mjs";
import { getDefaultSlots as n, validatePackage as a } from "@progress/kendo-vue-common";
const p = /* @__PURE__ */ e({
  name: "KendoHint",
  props: {
    id: String,
    direction: {
      type: String,
      default: "start",
      validator: function(t) {
        return ["start", "end"].includes(t);
      }
    }
  },
  created() {
    a(i);
  },
  computed: {
    hintClassName() {
      return {
        "k-form-hint": !0,
        "k-text-start": this.$props.direction === "start",
        "k-text-end": this.$props.direction === "end"
      };
    }
  },
  render() {
    const t = n(this);
    return r("div", {
      id: this.$props.id,
      class: this.hintClassName
    }, [t]);
  }
});
export {
  p as Hint
};
