// @progress/kendo-vue-data-tools/drag/DragClue.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2025 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
import { defineComponent as e, createVNode as t } from "vue";
import { Icon as s } from "@progress/kendo-vue-common";
import { cancelIcon as n, plusIcon as i } from "@progress/kendo-svg-icons";
const r = /* @__PURE__ */ e({
  data() {
    return {
      visible: !1,
      top: 0,
      left: 0,
      innerText: "",
      status: "cancel"
    };
  },
  render() {
    return this.visible && t("div", {
      class: "k-header k-drag-clue",
      style: {
        display: "block",
        position: "absolute",
        zIndex: 2e4,
        padding: "8px 12px",
        top: this.top + "px",
        left: this.left + "px"
      }
    }, [t(s, {
      name: this.status,
      class: "k-drag-status",
      icon: this.status === "cancel" ? n : i
    }, null), this.innerText]);
  }
});
export {
  r as DragClue
};
