// @progress/kendo-vue-inputs/textarea/TextArea.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2025 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
import { defineComponent as M, h as d, createVNode as t } from "vue";
import { getTabIndex as P, templateRendering as m, getListeners as g, getTemplate as x, Icon as l, isRtl as T, kendoThemeMaps as k, validatePackage as q, guid as D } from "@progress/kendo-vue-common";
import { packageMetadata as H } from "../package-metadata.mjs";
import { checkCircleIcon as L, exclamationCircleIcon as W, xIcon as F } from "@progress/kendo-svg-icons";
import { TextAreaResizeEnum as h } from "./interfaces/TextAreaResize.mjs";
const G = /* @__PURE__ */ M({
  name: "KendoTextArea",
  model: {
    event: "changemodel"
  },
  emits: {
    input: (e) => !0,
    change: (e) => !0,
    changemodel: (e) => !0,
    "update:modelValue": (e) => !0,
    focus: (e) => !0,
    blur: (e) => !0
  },
  props: {
    title: String,
    ariaDescribedBy: String,
    ariaLabelledBy: String,
    autoSize: Boolean,
    modelValue: {
      type: [String, Array, Number],
      default: void 0
    },
    defaultValue: [String, Number],
    dir: String,
    disabled: Boolean,
    readOnly: Boolean,
    rows: Number,
    id: String,
    name: String,
    validationMessage: String,
    size: {
      type: String,
      default: "medium",
      validator: function(e) {
        return [null, "small", "medium", "large"].includes(e);
      }
    },
    rounded: {
      type: String,
      default: "medium",
      validator: function(e) {
        return [null, "small", "medium", "large", "full"].includes(e);
      }
    },
    fillMode: {
      type: String,
      default: "solid",
      validator: function(e) {
        return [null, "solid", "flat", "outline"].includes(e);
      }
    },
    required: Boolean,
    placeholder: String,
    tabIndex: Number,
    valid: {
      type: Boolean,
      default: void 0
    },
    value: [String, Array, Number],
    validityStyles: {
      type: Boolean,
      default: !0
    },
    iconName: String,
    inputPrefix: [String, Function],
    inputSuffix: [String, Function],
    showValidationIcon: Boolean,
    showLoadingIcon: Boolean,
    showClearButton: Boolean,
    inputClass: String,
    wrapperClass: String,
    flow: {
      type: String,
      default: "horizontal",
      validator: function(e) {
        return ["horizontal", "vertical"].includes(e);
      }
    },
    resizable: {
      type: String,
      default: "vertical",
      validator: function(e) {
        return ["both", "horizontal", "vertical", "none"].includes(e);
      }
    },
    inputAttributes: Object
  },
  created() {
    q(H), this.calculatedId = D();
  },
  computed: {
    resizableClass() {
      const {
        resizable: e,
        autoSize: i
      } = this.$props;
      return i || e === h.none ? "k-resize-none" : e === h.vertical ? "k-resize-y" : e === h.horizontal ? "k-resize-x" : "k-resize";
    },
    rootClassName() {
      const {
        size: e,
        fillMode: i,
        rounded: s,
        required: a,
        showLoadingIcon: n
      } = this.$props, r = this.validityStyles === !0 ? !this.isValid : !1;
      return {
        "k-input": !0,
        "k-textarea": !0,
        [`k-input-${k.sizeMap[e] || e}`]: e,
        [`k-input-${i}`]: i,
        [`k-rounded-${k.roundedMap[s] || s}`]: s,
        "k-disabled": this.$props.disabled,
        "k-invalid": r,
        "k-required": a,
        "k-loading": n,
        [this.resizableClass]: !0,
        "!k-flex-col": this.flow === "vertical",
        "!k-flex-row": this.flow === "horizontal",
        [this.wrapperClass]: this.wrapperClass
      };
    },
    inputInnerClass() {
      return {
        "k-input-inner": !0,
        "!k-resize-none": !0,
        "!k-overflow-auto": !0,
        "k-flex": !0,
        [this.inputClass]: this.inputClass
      };
    },
    prefixRenderClass() {
      return {
        "k-input-prefix": !0,
        "!k-align-items-start": this.flow === "horizontal",
        "k-input-prefix-horizontal": this.flow === "vertical",
        "k-input-prefix-vertical": this.flow === "horizontal"
      };
    },
    suffixRenderClass() {
      return {
        "k-input-suffix": !0,
        "!k-align-items-start": this.flow === "horizontal",
        "k-input-suffix-horizontal": this.flow === "vertical",
        "k-input-suffix-vertical": this.flow === "horizontal"
      };
    },
    suffixIconWrapClass() {
      return {
        "k-flex-wrap": !0,
        "!k-align-items-start": !0
      };
    },
    computedValue() {
      let e;
      return this.$props.value !== void 0 ? e = this.$props.value : this.$props.modelValue !== void 0 ? e = this.$props.modelValue : e = this.currentValue, e;
    },
    isValid() {
      return this.valid !== void 0 ? this.valid : this.required ? !!this.computedValue : !0;
    }
  },
  data() {
    return {
      currentValue: this.$props.defaultValue,
      textAreaHeight: "auto",
      currentDir: "ltr"
    };
  },
  watch: {
    size() {
      this.textAreaHeight = "auto";
    },
    computedValue() {
      this.textAreaHeight = "auto";
    }
  },
  mounted() {
    this.element = this.elementRef;
    let e;
    this.$props.dir !== void 0 ? e = this.$props.dir : e = T(this.$el) ? "rtl" : "ltr", this.currentDir = e, this.setValidity();
  },
  updated() {
    this.element = this.elementRef, this.setValidity();
  },
  render() {
    const {
      ariaDescribedBy: e,
      ariaLabelledBy: i,
      autoSize: s,
      disabled: a,
      readOnly: n,
      required: r,
      rows: v,
      id: b,
      name: C,
      placeholder: V,
      tabIndex: y,
      iconName: o,
      showValidationIcon: u,
      showLoadingIcon: p,
      showClearButton: c,
      title: S,
      inputAttributes: $
    } = this.$props, w = {
      ...$,
      id: b || this.calculatedId,
      name: C,
      disabled: a,
      rows: v,
      title: S,
      placeholder: V,
      readOnly: n,
      required: r,
      tabIndex: P(y, a),
      "aria-labelledby": i,
      "aria-describedby": e,
      "aria-multiline": !0,
      "aria-disabled": a || void 0,
      ...this.$attrs
    }, z = m.call(this, this.$props.inputPrefix, g.call(this)), I = m.call(this, this.$props.inputSuffix, g.call(this)), B = x.call(this, {
      h: d,
      template: z,
      additionalProps: {
        value: this.computedValue,
        valid: this.isValid,
        className: this.prefixRenderClass
      }
    }), A = x.call(this, {
      h: d,
      template: I,
      additionalProps: {
        value: this.computedValue,
        valid: this.isValid,
        className: this.suffixRenderClass
      }
    }), N = d("textarea", {
      ...w,
      class: this.inputInnerClass,
      ref: (R) => {
        this.elementRef = R;
      },
      style: s ? {
        overflow: "hidden",
        height: this.textAreaHeight
      } : {},
      value: this.computedValue,
      onChange: this.handleChange,
      onInput: this.handleInput,
      onFocus: this.handleFocus,
      onBlur: this.handleBlur
    }), f = [o && t("span", {
      class: "k-flex-wrap"
    }, [t(l, {
      name: o,
      class: "k-input-icon"
    }, null)]), N, u && this.isValid && t("span", {
      class: this.suffixIconWrapClass
    }, [t(l, {
      name: "check",
      icon: L,
      class: "k-input-validation-icon"
    }, null)]), u && !this.isValid && t("span", {
      class: this.suffixIconWrapClass
    }, [t(l, {
      name: "exclamation-circle",
      icon: W,
      class: "k-input-validation-icon"
    }, null)]), p && t("span", {
      class: this.suffixIconWrapClass
    }, [t(l, {
      name: "loading",
      class: "k-input-loading-icon"
    }, null)]), c && this.computedValue && t("span", {
      class: this.suffixIconWrapClass
    }, [t("span", {
      onClick: this.clearClick,
      class: "k-clear-value"
    }, [t(l, {
      name: "x",
      icon: F
    }, null)])])];
    return t("span", {
      class: this.rootClassName,
      dir: this.currentDir === "rtl" ? this.currentDir : "",
      style: this.$attrs.style
    }, [this.$props.inputPrefix && B, o || u || p || c ? t("span", {
      class: "k-hstack k-flex"
    }, [f]) : f, this.$props.inputSuffix && A]);
  },
  methods: {
    setValidity() {
      var e, i;
      (i = (e = this.element) == null ? void 0 : e.setCustomValidity) == null || i.call(e, this.isValid ? "" : this.validationMessage || ""), this.element && (this.textAreaHeight = `${this.element.scrollHeight}px`);
    },
    clearClick(e) {
      this.emitUpdate(e, "change", "");
    },
    focus() {
      this.element && this.element.focus();
    },
    emitUpdate(e, i, s) {
      const a = s;
      this.$props.disabled || (this.currentValue = a), this.$props.disabled || (this.$emit("changemodel", a), this.$emit("update:modelValue", a), this.$emit(i, {
        event: e,
        component: this,
        name: this.element.name,
        value: a
      }));
    },
    handleChange(e) {
      this.emitUpdate(e, "change", e.target.value);
    },
    handleInput(e) {
      this.emitUpdate(e, "input", e.target.value);
    },
    handleFocus(e) {
      this.$props.disabled || this.$emit("focus", {
        event: e,
        component: this,
        name: this.element.name
      });
    },
    handleBlur(e) {
      this.$props.disabled || this.$emit("blur", {
        event: e,
        component: this,
        name: this.element.name
      });
    }
  }
});
export {
  G as TextArea
};
