// @progress/kendo-vue-grid/footer/FooterRow.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2025 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
import { defineComponent as i, createVNode as l, h as d } from "vue";
import { getTemplate as n } from "@progress/kendo-vue-common";
import { footerColumns as p } from "../utils/main.mjs";
const g = /* @__PURE__ */ i({
  name: "FooterRow",
  props: {
    isRtl: Boolean,
    columns: Array,
    rowIndex: Number
  },
  methods: {
    columnStyles(t) {
      return t.left !== void 0 ? this.$props.isRtl ? {
        left: t.right + "px",
        right: t.left + "px"
      } : {
        left: t.left + "px",
        right: t.right + "px"
      } : {};
    }
  },
  render() {
    const t = function(e, s) {
      let r = null;
      const o = e.locked && e.left !== void 0 ? "k-grid-footer-sticky" : "", a = e.footerClassName ? `k-table-td ${e.footerClassName} ${o}` : "k-table-td " + o;
      return r = n.call(this, {
        h: d,
        template: e.footerCell,
        defaultRendering: null,
        additionalProps: {
          field: e.field,
          colSpan: e.colSpan !== 1 ? e.colSpan : void 0,
          defaultStyle: this.columnStyles(e)
        }
      }), l("td", {
        key: s,
        colspan: e.colSpan !== 1 ? e.colSpan : void 0,
        style: this.columnStyles(e),
        class: a,
        role: "gridcell"
      }, [r]);
    };
    return l("tr", {
      class: "k-table-row",
      role: "row",
      "aria-rowIndex": this.$props.rowIndex
    }, [p(this.$props.columns).map(t, this)]);
  }
});
export {
  g as FooterRow
};
