// @progress/kendo-vue-data-tools/pager/PagerNavigationButton.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2025 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
import { defineComponent as t, createVNode as n } from "vue";
import { Button as i } from "@progress/kendo-vue-buttons";
import { getIconName as o } from "@progress/kendo-vue-common";
const s = /* @__PURE__ */ t({
  name: "KendoPagerNavigationButton",
  props: {
    title: String,
    disabled: Boolean,
    icon: String,
    svgIcon: Object,
    page: Number,
    size: String,
    onPagechange: Function
  },
  inject: {
    kendoLocalizationService: {
      default: null
    },
    kendoIntlService: {
      default: null
    }
  },
  methods: {
    changePage(e) {
      e.preventDefault(), this.$emit("pagechange", this.$props.page, e);
    }
  },
  render() {
    return n(i, {
      fillMode: "flat",
      themeColor: "base",
      size: this.$props.size,
      rounded: null,
      class: "k-pager-nav",
      ariaLabel: this.title,
      icon: o(this.icon),
      svgIcon: this.svgIcon,
      title: this.title,
      disabled: this.disabled,
      role: "button",
      onClick: this.changePage
    }, null);
  }
});
export {
  s as PagerNavigationButton
};
