// @progress/kendo-vue-data-tools/pager/PagerInfo.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2025 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
import { defineComponent as i, createVNode as n, inject as a } from "vue";
import { provideIntlService as r, provideLocalizationService as p } from "@progress/kendo-vue-intl";
import { pagerInfo as t, messages as c } from "../messages/main.mjs";
const d = /* @__PURE__ */ i({
  name: "KendoPagerInfo",
  props: {
    totalPages: Number,
    currentPage: Number,
    skip: Number,
    messagesMap: Function
  },
  inject: {
    kendoLocalizationService: {
      default: null
    },
    kendoIntlService: {
      default: null
    }
  },
  setup() {
    const e = a("kendoIntlService", {}), s = a("kendoLocalizationService", {});
    return {
      kendoIntlService: e,
      kendoLocalizationService: s
    };
  },
  render() {
    const e = r(this), s = p(this), o = this.$props.messagesMap ? this.$props.messagesMap(t) : {
      messageKey: t,
      defaultMessage: c[t]
    };
    return n("span", {
      class: "k-pager-info"
    }, [e.format(s.toLanguageString(o.messageKey, o.defaultMessage), [Math.min(this.$props.skip + 1, this.$props.totalPages), Math.min(this.$props.skip + this.$props.currentPage, this.$props.totalPages), this.$props.totalPages])]);
  }
});
export {
  d as PagerInfo
};
