// @progress/kendo-vue-grid/GridToolbar.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2025 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
import { defineComponent as o, createVNode as e } from "vue";
import { getDefaultSlots as t, kendoThemeMaps as a } from "@progress/kendo-vue-common";
const i = /* @__PURE__ */ o({
  name: "GridToolbar",
  props: {
    size: {
      type: String,
      default: "medium",
      validator: function(r) {
        return ["small", "medium"].includes(r);
      }
    }
  },
  computed: {
    wrapperClass() {
      const {
        size: r
      } = this.$props;
      return {
        "k-toolbar": !0,
        "k-grid-toolbar": !0,
        "k-toolbar-solid": !0,
        [`k-toolbar-${a.sizeMap[r] || r}`]: r
      };
    }
  },
  render() {
    const r = t(this);
    return e("div", {
      class: this.wrapperClass,
      role: "toolbar"
    }, [r]);
  }
});
export {
  i as GridToolbar
};
