// @progress/kendo-vue-grid/GridNoRecords.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2025 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
import { defineComponent as t, createVNode as n, inject as r } from "vue";
import { provideLocalizationService as c } from "@progress/kendo-vue-intl";
import { noRecords as o, messages as a } from "./messages/main.mjs";
import { getDefaultSlots as l } from "@progress/kendo-vue-common";
const v = /* @__PURE__ */ t({
  name: "GridNoRecords",
  inject: {
    kendoLocalizationService: {
      default: null
    }
  },
  setup() {
    return {
      kendoLocalizationService: r("kendoLocalizationService", {})
    };
  },
  render() {
    const e = l(this), i = c(this).toLanguageString(o, a[o]);
    return e ? n("div", null, [e]) : i;
  }
});
export {
  v as GridNoRecords
};
