// @progress/kendo-vue-inputs/colors/common/ColorsAdaptiveMode.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2025 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
import { defineComponent as h, h as l, createVNode as t, Fragment as f, isVNode as u } from "vue";
import { ActionSheet as y } from "@progress/kendo-vue-layout";
import { Button as a } from "@progress/kendo-vue-buttons";
import { xIcon as k } from "@progress/kendo-svg-icons";
function i(e) {
  return typeof e == "function" || Object.prototype.toString.call(e) === "[object Object]" && !u(e);
}
const A = /* @__PURE__ */ h({
  name: "KendoActionSheet",
  props: {
    expand: Boolean,
    animationStyles: Object,
    classNameAdaptive: String,
    title: String,
    componentToRender: Object,
    applyText: String,
    cancelText: String
  },
  emits: ["actionSheetClose", "cancelBtnClick", "applyBtnClick"],
  methods: {
    onClose() {
      this.$emit("actionSheetClose");
    },
    onCancel() {
      this.$emit("cancelBtnClick");
    },
    onApply(e) {
      this.$emit("applyBtnClick", e);
    }
  },
  render() {
    const {
      title: e,
      componentToRender: c,
      expand: r,
      applyText: o,
      cancelText: n,
      animationStyles: s,
      classNameAdaptive: p
    } = this.$props, d = function() {
      return l(t(a, {
        tabIndex: 0,
        size: "large",
        "aria-label": "Cancel",
        "aria-disabled": "false",
        type: "button",
        fillMode: "flat",
        icon: "x",
        svgIcon: k,
        onClick: this.onClose
      }, null));
    }.call(this), m = () => l(t(f, null, [t(a, {
      size: "large",
      class: "k-coloreditor-cancel",
      "aria-label": n,
      onClick: this.onCancel
    }, i(n) ? n : {
      default: () => [n]
    }), t(a, {
      themeColor: "primary",
      size: "large",
      class: "k-coloreditor-apply",
      "aria-label": o,
      onClick: this.onApply
    }, i(o) ? o : {
      default: () => [o]
    })])), C = () => l(c);
    return t(y, {
      onClose: this.onClose,
      animationStyles: s,
      className: p,
      contentClassName: "!k-overflow-hidden",
      footerClassName: "k-actions k-actions-stretched k-actions-horizontal",
      title: e,
      content: C,
      footer: m,
      expand: r,
      actions: d
    }, null);
  }
});
export {
  A as ColorsAdaptiveMode
};
