// @progress/kendo-vue-inputs/colors/HexInput.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2025 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
import { defineComponent as s, createVNode as o } from "vue";
import { parseColor as i } from "./utils/color-parser.mjs";
import { isPresent as n } from "./utils/misc.mjs";
import { TextBox as h } from "../textbox/TextBox.mjs";
const d = /* @__PURE__ */ s({
  name: "KendoHexInput",
  emits: {
    hexchange: (e) => !0,
    blur: (e) => !0,
    focus: (e) => !0
  },
  props: {
    tabIndex: Number,
    hex: String,
    disabled: Boolean,
    id: String,
    size: String
  },
  computed: {
    isHexValid() {
      return !!i(this.currentHex, "rgba");
    }
  },
  data() {
    return {
      currentHex: this.$props.hex,
      originalHex: this.$props.hex
    };
  },
  mounted() {
    this._input = this.$refs.input._input;
  },
  watch: {
    hex: function(e) {
      this.currentHex = e;
    }
  },
  render() {
    return o(h, {
      id: this.$props.id,
      value: this.currentHex,
      onInput: this.onChange,
      onChange: this.onChange,
      onFocus: this.onFocus,
      onBlur: this.onBlur,
      valid: this.isHexValid,
      disabled: this.$props.disabled,
      size: this.$props.size,
      tabIndex: this.tabIndex,
      ref: "input"
    }, null);
  },
  methods: {
    onChange(e) {
      const t = e.target.value, r = i(t, "rgba");
      this.currentHex = t, n(r) && this.$emit("hexchange", {
        hex: t,
        value: r,
        event: e
      });
    },
    onBlur(e) {
      n(i(this.hex, "rgba")) || (this.currentHex = this.originalHex), this.$emit("blur", {
        event: e
      });
    },
    onFocus(e) {
      this.$emit("focus", {
        event: e
      });
    }
  }
});
export {
  d as HexInput
};
