// @progress/kendo-vue-layout/menu/components/MenuItemLink.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2025 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
import { defineComponent as n, createVNode as t } from "vue";
import { getDefaultSlots as o } from "@progress/kendo-vue-common";
const a = /* @__PURE__ */ n({
  name: "KendoMenuItemLink",
  props: {
    opened: Boolean,
    focused: Boolean,
    url: String
  },
  computed: {
    menuItemClassName() {
      return {
        "k-link": !0,
        "k-menu-link": !0,
        "k-active": this.$props.opened,
        "k-focus": this.$props.focused
      };
    }
  },
  render() {
    const e = o(this);
    return this.$props.url ? t("a", {
      class: this.menuItemClassName,
      role: "presentation",
      href: this.$props.url,
      tabindex: -1
    }, [e]) : t("span", {
      class: this.menuItemClassName,
      role: "presentation"
    }, [e]);
  }
});
export {
  a as MenuItemLink
};
