// @progress/kendo-vue-layout/menu/components/MenuItemArrow.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2025 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
import { defineComponent as t, createVNode as o } from "vue";
import { Icon as e } from "@progress/kendo-vue-common";
import { caretAltDownIcon as n, caretAltRightIcon as i, caretAltLeftIcon as c } from "@progress/kendo-svg-icons";
import { getChildrenPosition as a } from "../utils/misc.mjs";
const s = n, d = i, m = c, I = /* @__PURE__ */ t({
  name: "KendoMenuItemArrow",
  props: {
    itemId: String,
    dir: String,
    verticalMenu: Boolean
  },
  render() {
    return o("span", {
      class: "k-menu-expand-arrow",
      "aria-hidden": "true"
    }, [o(e, {
      icon: this.getArrowName(),
      role: "presentation"
    }, null)]);
  },
  methods: {
    getArrowName() {
      const r = a(this.$props.itemId, this.$props.verticalMenu === !0, this.$props.dir === "rtl");
      return r === "downward" ? s : r === "rightward" ? d : r === "leftward" ? m : "";
    }
  }
});
export {
  I as MenuItemArrow,
  s as downArrowIcon,
  m as leftArrowIcon,
  d as rightArrowIcon
};
