// @progress/kendo-vue-excel-export/ooxml/ExporterColumns.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2025 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
const d = (t, e, r) => (p) => {
  r(e, p);
  const o = t({
    ...e
  }), l = Number(o);
  return isNaN(l) ? o : l;
}, g = (t, e) => {
  t.$implicit = t.group = e, t.field = e.field, t.value = e.value, t.aggregates = e.aggregates;
}, n = (t, e) => {
  t.group = e.group, t.$implicit = t.aggregates = e;
}, u = (t, e) => {
}, C = (t) => {
  const e = [];
  let r = 0;
  const s = (p, o, l) => {
    p.forEach((i) => {
      const h = new f(i, r);
      if (o.push(h), i.children && i.children.length > 1) {
        const a = h.columns = [];
        s(i.children, a);
      } else
        r++;
    });
  };
  return s(t, e), e;
};
class f {
  constructor(e, r) {
    this.title = e.title, this.field = e.field, this.hidden = e.hidden, this.locked = e.locked, this.width = e.width, this.headerCellOptions = e.headerCellOptions, this.cellOptions = e.cellOptions, this.groupHeaderCellOptions = e.groupHeaderCellOptions, this.groupFooterCellOptions = e.groupFooterCellOptions, this.footerCellOptions = e.footerCellOptions, e.footer && (this.footerTemplate = d(e.footer, {
      $implicit: e,
      column: e,
      columnIndex: r
    }, u)), e.groupFooter && (this.groupFooterTemplate = d(e.groupFooter, {
      column: e,
      field: e.field
    }, n)), e.groupHeader && (this.groupHeaderTemplate = d(e.groupHeader, {}, g));
  }
}
export {
  f as ExporterColumn,
  C as toExporterColumns
};
