// @progress/kendo-vue-grid/cells/GridDetailHierarchyCell.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2025 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
import { defineComponent as t, createVNode as r, inject as n } from "vue";
import { KEYBOARD_NAV_DATA_ID as i, KEYBOARD_NAV_DATA_LEVEL as o } from "@progress/kendo-vue-data-tools";
import { noop as a } from "@progress/kendo-vue-common";
const s = /* @__PURE__ */ t({
  props: {
    id: String
  },
  inject: {
    getKeyboardNavigationAttributes: {
      default: a
    }
  },
  setup() {
    return {
      kendoIntlService: n("kendoIntlService", {})
    };
  },
  render() {
    const e = this.getKeyboardNavigationAttributes(this.$props.id);
    return r("td", {
      class: "k-table-td k-hierarchy-cell",
      tabindex: e.tabIndex,
      "data-keyboardnavlevel": e[o],
      "data-keyboardnavid": e[i]
    }, null);
  }
});
export {
  s as GridDetailHierarchyCell
};
