// @progress/kendo-vue-grid/cells/GridDetailCell.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2025 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
import { defineComponent as l, createVNode as c, inject as s, h as p } from "vue";
import { noop as m, getTemplate as b } from "@progress/kendo-vue-common";
import { KEYBOARD_NAV_DATA_ID as u, KEYBOARD_NAV_DATA_LEVEL as I } from "@progress/kendo-vue-data-tools";
const x = /* @__PURE__ */ l({
  props: {
    colSpan: Number,
    ariaColIndex: Number,
    dataItem: [Object, String, Number],
    dataIndex: Number,
    detail: [String, Function, Object],
    id: String
  },
  inject: {
    getKeyboardNavigationAttributes: {
      default: m
    }
  },
  setup() {
    return {
      kendoIntlService: s("kendoIntlService", {})
    };
  },
  render() {
    const {
      colSpan: e,
      ariaColIndex: a,
      dataItem: n,
      dataIndex: r,
      id: i
    } = this.$props, t = this.getKeyboardNavigationAttributes(i), d = function(o) {
      return b.call(this, {
        h: p,
        template: this.$props.detail,
        additionalProps: o
      });
    };
    return c("td", {
      class: "k-table-td k-detail-cell",
      colspan: e,
      "aria-colindex": a,
      role: "gridcell",
      tabindex: t.tabIndex,
      "data-keyboardnavlevel": t[I],
      "data-keyboardnavid": t[u]
    }, [d.call(this, {
      dataItem: n,
      dataIndex: r
    })]);
  }
});
export {
  x as GridDetailCell
};
