// @progress/kendo-vue-dialogs/Dialog.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2025 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
import { defineComponent as I, createVNode as i } from "vue";
import { DialogTitleBar as C } from "./DialogTitleBar.mjs";
import { templateRendering as k, getListeners as y, getDefaultSlots as S, WatermarkOverlay as D, guid as h, Keys as $, hasListener as x, validatePackage as b, shouldShowValidationUI as v, getLicenseMessage as L } from "@progress/kendo-vue-common";
import { packageMetadata as d } from "./package-metadata.mjs";
import { DEFAULT_DIALOGS_ZINDEX as T } from "./constants.mjs";
const K = /* @__PURE__ */ I({
  name: "KendoDialog",
  emits: {
    close: (e) => !0,
    overlayclick: null
  },
  props: {
    appendTo: String,
    title: String,
    titleRender: [String, Function],
    id: String,
    wrapperId: String,
    dir: String,
    closeIcon: {
      type: Boolean,
      default: !0
    },
    width: [String, Number],
    height: [String, Number],
    minWidth: [String, Number],
    className: String,
    dialogClass: String,
    themeColor: {
      type: String,
      validator: function(e) {
        return ["primary", "dark", "light"].includes(e);
      }
    },
    onClose: Function
  },
  provide() {
    return {
      kCurrentZIndex: T
    };
  },
  data() {
    return {
      showLicenseWatermark: !1,
      licenseMessage: void 0
    };
  },
  created() {
    b(d), this.showLicenseWatermark = v(d), this.licenseMessage = L(d), this.titleId = this.generateTitleId(), this.contentId = this.generateContentId();
  },
  computed: {
    wrapperClass() {
      const {
        className: e
      } = this.$props;
      return {
        "k-dialog-wrapper": !0,
        [e]: e
      };
    },
    dialogElementClass() {
      const {
        dialogClass: e,
        themeColor: n
      } = this.$props;
      return {
        "k-window": !0,
        "k-dialog": !0,
        [e]: e,
        [`k-window-${n}`]: n
      };
    }
  },
  mounted() {
    if (this.$props.appendTo) {
      const e = document.querySelector(this.$props.appendTo);
      this.windowElement = this.$refs.wrapper, e.append(this.windowElement);
    }
  },
  beforeUnmount() {
    this.$props.appendTo && this.windowElement.remove();
  },
  methods: {
    handleCloseDialog(e) {
      e.preventDefault(), this.$emit("close", {
        event: e,
        target: this
      });
    },
    handleKeyDown(e) {
      e.keyCode === $.esc && x.call(this, "close") && (e.preventDefault(), this.handleCloseDialog(e));
    },
    transformDimesion(e) {
      return typeof e == "string" && (e.endsWith("px") || e.endsWith("%")) ? e : e + "px";
    },
    getActionBarIndex(e) {
      return e.findIndex((t) => t && t.tag && t.tag.toLowerCase().indexOf("dialogactionsbar") !== -1 || t.componentOptions && t.componentOptions.tag && t.componentOptions.tag.toLowerCase().indexOf("actions-bar") !== -1 || t.type && t.type.name && t.type.name.toLowerCase().indexOf("dialogactionsbar") !== -1);
    },
    generateTitleId() {
      return "dialog-title" + h();
    },
    generateContentId() {
      return "dialog-content" + h();
    },
    handleClick(e) {
      this.$emit("overlayclick", e);
    }
  },
  render() {
    const e = this.$props.id !== void 0 ? this.$props.id : this.titleId;
    let {
      title: n,
      width: t,
      height: s,
      minWidth: r,
      dir: g,
      contentStyle: m,
      wrapperId: u
    } = this.$props;
    const a = this.$props.titleRender ? k.call(this, this.$props.titleRender, y.call(this)) : null, o = S(this) || [];
    t = this.transformDimesion(t), s = this.transformDimesion(s), r = this.transformDimesion(r);
    const f = this.showLicenseWatermark ? i(D, {
      message: this.licenseMessage
    }, null) : null, l = this.getActionBarIndex(o);
    let p;
    l !== -1 && (p = o[l], o.splice(l, 1));
    const w = this.$props.closeIcon !== void 0 ? this.$props.closeIcon : !0, c = i("div", {
      ref: "wrapper",
      class: this.wrapperClass,
      onKeydown: this.handleKeyDown,
      tabindex: 0,
      id: u,
      dir: g
    }, [i("div", {
      class: "k-overlay",
      onClick: this.handleClick
    }, null), i("div", {
      "aria-labelledby": n || a ? e : void 0,
      "aria-describedby": this.contentId,
      "aria-modal": !0,
      class: this.dialogElementClass,
      role: "dialog",
      style: {
        width: t,
        height: s,
        minWidth: r
      }
    }, [(n || a) && i(C, {
      closeIcon: w,
      onClosebuttonclick: this.handleCloseDialog,
      id: e,
      title: n,
      titleRender: a
    }, null), i("div", {
      class: "k-window-content k-dialog-content",
      style: m,
      id: this.contentId
    }, [o]), p, f])]);
    return this.$props.appendTo ? i("div", null, [c]) : c;
  }
});
export {
  K as Dialog
};
