// @progress/kendo-vue-dialogs/DialogTitleBar.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2025 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
import { defineComponent as a, h as r, createVNode as t } from "vue";
import { getTemplate as c } from "@progress/kendo-vue-common";
import { Button as s } from "@progress/kendo-vue-buttons";
import { xIcon as d } from "@progress/kendo-svg-icons";
const b = /* @__PURE__ */ a({
  props: {
    id: String,
    closeIcon: {
      type: Boolean,
      default: !0
    },
    title: String,
    titleRender: [String, Function, Object],
    onClosebuttonclick: Function
  },
  methods: {
    onCloseButtonClick(e) {
      this.$emit("closebuttonclick", e);
    }
  },
  render() {
    let {
      id: e,
      closeIcon: o,
      titleRender: l,
      title: n
    } = this.$props, i;
    return i = c.call(this, {
      h: r,
      template: l,
      defaultRendering: n
    }), t("div", {
      class: "k-window-titlebar k-dialog-titlebar",
      id: e
    }, [t("span", {
      class: "k-window-title k-dialog-title"
    }, [i]), t("div", {
      class: "k-window-titlebar-actions k-dialog-titlebar-actions"
    }, [o && t(s, {
      type: "button",
      fillMode: "flat",
      "aria-label": "Close",
      icon: "x",
      svgIcon: d,
      onClick: this.onCloseButtonClick,
      class: "k-window-titlebar-action k-dialog-titlebar-action"
    }, null)])]);
  }
});
export {
  b as DialogTitleBar
};
