// @progress/kendo-vue-dialogs/DialogActionsBar.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2025 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
import { defineComponent as o, createVNode as r } from "vue";
import { getDefaultSlots as n } from "@progress/kendo-vue-common";
const s = /* @__PURE__ */ o({
  name: "DialogActionsBar",
  props: {
    layout: {
      type: String,
      default: "stretched",
      validator: function(t) {
        return ["stretched", "start", "center", "end"].includes(t);
      }
    },
    orientation: {
      type: String,
      default: "horizontal",
      validator: function(t) {
        return ["horizontal", "vertical"].includes(t);
      }
    }
  },
  computed: {
    wrapperClasses() {
      const {
        layout: t,
        orientation: e
      } = this.$props;
      return {
        "k-actions": !0,
        "k-dialog-actions": !0,
        "k-window-actions": !0,
        [`k-actions-${e}`]: e,
        [`k-actions-${t}`]: t
      };
    }
  },
  render() {
    const t = n(this);
    return r("div", {
      class: this.wrapperClasses
    }, [t]);
  }
});
export {
  s as DialogActionsBar
};
