// @progress/kendo-vue-dialogs/WindowResizeHandlers.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2025 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
import { defineComponent as s, createVNode as r } from "vue";
import { Draggable as t } from "@progress/kendo-vue-common";
const o = ["n", "e", "s", "w", "se", "sw", "ne", "nw"], d = /* @__PURE__ */ s({
  name: "ResizeHandlers",
  emits: {
    resize: null
  },
  methods: {
    onDrag(e, n) {
      e.originalEvent.preventDefault(), this.$emit("resize", e, {
        end: !1,
        direction: n
      });
    },
    onRelease(e, n) {
      e.originalEvent.preventDefault(), this.$emit("resize", e, {
        end: !0,
        direction: n
      });
    }
  },
  render() {
    return r("div", null, [o.map(function(e, n) {
      return r(t, {
        key: n,
        onDrag: (i) => this.onDrag(i, e),
        onRelease: (i) => this.onRelease(i, e)
      }, {
        default: () => [r("div", {
          class: "k-resize-handle k-resize-" + e,
          style: {
            display: "block",
            touchAction: "none"
          }
        }, null)]
      });
    }, this)]);
  }
});
export {
  d as ResizeHandlers
};
