// @progress/kendo-vue-treeview/utils/FieldsService.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2025 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
var h = Object.defineProperty;
var c = (t, e, l) => e in t ? h(t, e, { enumerable: !0, configurable: !0, writable: !0, value: l }) : t[e] = l;
var d = (t, e, l) => c(t, typeof e != "symbol" ? e + "" : e, l);
import { getNestedValue as i } from "./misc.mjs";
class F {
  constructor(e) {
    d(this, "focusIdField");
    d(this, "expandField");
    d(this, "selectField");
    d(this, "childrenField");
    d(this, "hasChildrenField");
    d(this, "textField");
    d(this, "disableField");
    d(this, "checkField");
    d(this, "checkIndeterminateField");
    this.expandField = e.expandField, this.selectField = e.selectField, this.hasChildrenField = e.hasChildrenField, this.childrenField = e.childrenField, this.textField = e.textField, this.disableField = e.disableField, this.checkField = e.checkField, this.checkIndeterminateField = e.checkIndeterminateField, this.focusIdField = e.focusIdField;
  }
  expanded(e) {
    return i(this.expandField, e);
  }
  selected(e) {
    return i(this.selectField, e);
  }
  text(e) {
    return i(this.textField, e);
  }
  disabled(e) {
    return i(this.disableField, e);
  }
  hasChildren(e) {
    return i(this.hasChildrenField, e);
  }
  children(e) {
    const l = i(this.childrenField, e);
    return l || [];
  }
  checked(e) {
    return i(this.checkField, e);
  }
  checkIndeterminate(e) {
    return i(this.checkIndeterminateField, e);
  }
  focusId(e) {
    return this.focusIdField && i(this.focusIdField, e);
  }
  getChildrenField() {
    return this.childrenField;
  }
}
export {
  F as default
};
