// @progress/kendo-vue-treeview/utils/getItemIdUponKeyboardNavigation.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2025 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
import { Keys as l } from "@progress/kendo-vue-common";
import { ZERO_LEVEL_ZERO_NODE_ID as I, getDirectParentId as i, isIdZeroLevel as E, getFirstChildId as y, getShortId as C, createId as p, getItemById as s } from "./itemIdUtils.mjs";
import { isEnabledAndAllParentsEnabled as N, isItemExpandedAndWithChildren as d } from "./itemUtils.mjs";
function D(r, n, e, o, t) {
  switch (o) {
    case l.left:
      return c();
    case l.right:
      return a();
    case l.up:
      return b();
    case l.down:
      return x();
    case l.home:
      return I;
    case l.end:
      return F();
    default:
      return n;
  }
  function c() {
    return E(n) || t.expanded(r) && N(n, e, t) ? n : i(n);
  }
  function a() {
    return d(r, t) ? y(n) : n;
  }
  function b() {
    const u = Number(C(n)), g = i(n);
    return u ? U(p(u - 1, g), e, t) : E(n) ? n : g;
  }
  function x() {
    return d(r, t) ? y(n) : K(n, e, t) || n;
  }
  function F() {
    let u = (e.length - 1).toString(), g = e[e.length - 1], h;
    for (; d(g, t); )
      h = g[t.getChildrenField()], u = p(h.length - 1, u), g = h[h.length - 1];
    return u;
  }
}
function K(r, n, e) {
  const o = i(r), t = o ? s(o, n, e.getChildrenField()) : void 0, c = t ? t[e.getChildrenField()] : n, a = Number(C(r));
  return a < c.length - 1 ? p(a + 1, o) : t ? K(o, n, e) : void 0;
}
function U(r, n, e) {
  const o = s(r, n, e.getChildrenField());
  return d(o, e) ? U(
    p(
      o[e.getChildrenField()].length - 1,
      r
    ),
    n,
    e
  ) : r;
}
export {
  D as default
};
