// @progress/kendo-vue-treeview/utils/itemIdUtils.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2025 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
const c = "", x = "0", r = "_";
function l(t, n, s) {
  if (O(t))
    return n[Number(t)];
  {
    const e = n[Number(f(t))], o = e && e[s] || [];
    return o.length ? l(a(t), o, s) : void 0;
  }
}
function f(t) {
  return g(t) ? t : t.split(r)[0];
}
function a(t) {
  if (g(t))
    return t;
  {
    const n = t.indexOf(r);
    return t.substring(n + 1);
  }
}
function E(t) {
  return b("0", t);
}
function b(t, n) {
  return t = t.toString(), n ? n + r + t : t;
}
function S(t) {
  const n = t.lastIndexOf(r);
  return n < 0 ? c : t.substring(0, n);
}
function g(t) {
  return t === c || t.indexOf(r) < 0;
}
function O(t) {
  return t !== c && t.indexOf(r) < 0;
}
function P(t) {
  return t.split(r);
}
function h(t) {
  const n = t.lastIndexOf(r);
  return n < 0 ? t : t.substring(n + 1);
}
function p(t, n) {
  const s = n;
  t = "r" + r + t, n = "r" + r + n;
  const e = S(t) + r;
  if (n.startsWith(e)) {
    const o = n.substring(e.length);
    if (o) {
      const u = f(o);
      if (Number(h(t)) < Number(u)) {
        const i = e + (Number(u) - 1).toString() + o.substring(u.length);
        return a(i);
      }
    }
  }
  return s;
}
export {
  c as EMPTY_ID,
  r as SEPARATOR,
  x as ZERO_LEVEL_ZERO_NODE_ID,
  b as createId,
  P as getAllShortIds,
  p as getDecrementedItemIdAfterRemoval,
  S as getDirectParentId,
  E as getFirstChildId,
  a as getIdWithoutRootParentId,
  l as getItemById,
  f as getRootParentId,
  h as getShortId,
  g as isIdEmptyOrZeroLevel,
  O as isIdZeroLevel
};
