// @progress/kendo-vue-inputs/messages/main.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2025 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
const e = "numerictextbox.increment", o = "numerictextbox.decrement", t = "slider.increment", r = "slider.decrement", a = "slider.dragTitle", n = "colorGradient.dragHandle", l = "colorGradient.r", i = "colorGradient.g", c = "colorGradient.b", s = "colorGradient.a", d = "colorGradient.hex", G = "colorGradient.hexLabel", u = "colorGradient.inputColorButton", g = "colorGradient.rLabel", b = "colorGradient.gLabel", p = "colorGradient.bLabel", h = "colorGradient.aLabel", A = "colorGradient.contrastRatio", C = "colorGradient.colorGradientAALevel", L = "colorGradient.colorGradientAAALevel", P = "colorGradient.colorGradientPass", x = "colorGradient.colorGradientFail", B = "colorGradient.hueSliderLabel", k = "colorGradient.alphaSliderLabel", m = "flatColorPicker.cancelBtn", f = "flatColorPicker.applyBtn", v = "flatColorPicker.gradientBtn", w = "flatColorPicker.paletterBtn", D = "flatColorPicker.clearBtn", T = "colorPicker.adaptiveTitle", H = "checkbox.validation", I = "checkbox.optionalText", V = "radioButton.validation", y = "switch.validation", z = "colorPicker.dropdownButtonAriaLabel", S = "colorGradient.toggleInputsButton", R = "rating.ariaLabel", M = "signature.clear", F = "signature.maximize", O = "signature.minimize", E = {
  [e]: "Increase value",
  [o]: "Decrease value",
  [t]: "Increase",
  [r]: "Decrease",
  [a]: "Drag",
  [n]: `Color well with two-dimensional slider for selecting
     saturation and lightness. Selected color is`,
  [u]: "Change the color input",
  [l]: "r",
  [i]: "g",
  [c]: "b",
  [s]: "a",
  [g]: "red chanel",
  [b]: "green chanel",
  [p]: "blue chanel",
  [h]: "alpha chanel",
  [d]: "hex",
  [G]: "HEX",
  [A]: "Contrast ratio",
  [C]: "AA",
  [L]: "AAA",
  [P]: "Pass",
  [x]: "Fail",
  [B]: "Hue slider",
  [k]: "Alpha slider",
  [S]: "Toggle colorgradient inputs",
  [m]: "Cancel",
  [f]: "Apply",
  [v]: "Gradient",
  [w]: "Palette",
  [D]: "Clear",
  [T]: "Choose color",
  [H]: "Please check this box if you want to proceed!",
  [I]: "(Optional)",
  [V]: "Please select option if you want to proceed!",
  [y]: "Please turn on if you want to proceed!",
  [z]: "Select",
  [R]: "Rating",
  [M]: "Clear",
  [F]: "Maximize",
  [O]: "Minimize"
};
export {
  I as checkboxOptionalText,
  H as checkboxValidation,
  s as colorGradientA,
  L as colorGradientAAALevel,
  C as colorGradientAALevel,
  h as colorGradientALabel,
  k as colorGradientAlphaSliderLabel,
  c as colorGradientB,
  p as colorGradientBLabel,
  A as colorGradientContrastRatio,
  n as colorGradientDragHandle,
  x as colorGradientFail,
  i as colorGradientG,
  b as colorGradientGLabel,
  d as colorGradientHex,
  G as colorGradientHexLabel,
  B as colorGradientHueSliderLabel,
  u as colorGradientInputColorButton,
  P as colorGradientPass,
  l as colorGradientR,
  g as colorGradientRLabel,
  S as colorGradientToggleInputsButton,
  T as colorPickerAdaptiveTitle,
  z as colorPickerDropdownButtonAriaLabel,
  f as flatColorPickerApplyBtn,
  m as flatColorPickerCancelBtn,
  D as flatColorPickerClearBtn,
  v as flatColorPickerGradientBtn,
  w as flatColorPickerPaletteBtn,
  E as messages,
  o as numericDecreaseValue,
  e as numericIncreaseValue,
  V as radioButtonValidation,
  R as ratingAriaLabel,
  M as signatureClear,
  F as signatureMaximize,
  O as signatureMinimize,
  r as sliderDecreaseValue,
  a as sliderDragTitle,
  t as sliderIncreaseValue,
  y as switchValidation
};
