// @progress/kendo-vue-progressbars/progressbar/ProgressBar.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2025 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
import { defineComponent as P, h as m, createVNode as r } from "vue";
import { animate as g, cancelAnimation as x } from "@progress/kendo-vue-animation";
import { templateRendering as R, getListeners as D, getTemplate as C, classNames as E, getTabIndex as I, setRef as b, getRef as f, isRtl as O, validatePackage as W } from "@progress/kendo-vue-common";
import { calculatePercentage as o, updateProgress as l, truncateNumber as A } from "../common/utils.mjs";
import { DEFAULT_ANIMATION_DURATION as B, NO_ANIMATION as U } from "../common/constants.mjs";
import { packageMetadata as w } from "../package-metadata.mjs";
const K = /* @__PURE__ */ P({
  name: "KendoProgressBar",
  props: {
    animation: {
      type: [Boolean, Object],
      default: !1
    },
    disabled: {
      type: Boolean,
      default: !1
    },
    reverse: {
      type: Boolean,
      default: !1
    },
    label: String,
    labelRender: [String, Object, Function],
    labelVisible: {
      type: Boolean,
      default: !0
    },
    labelPlacement: {
      type: String,
      default: void 0,
      validator: function(e) {
        return ["start", "center", "end"].includes(e);
      }
    },
    dir: {
      type: String,
      default: void 0
    },
    max: {
      type: Number,
      default: 100
    },
    min: {
      type: Number,
      default: 0
    },
    value: {
      type: Number,
      default: 0
    },
    orientation: {
      type: String,
      default: "horizontal",
      validator: function(e) {
        return ["horizontal", "vertical"].includes(e);
      }
    },
    tabIndex: Number,
    emptyStyle: Object,
    emptyClassName: String,
    progressStyle: Object,
    progressClassName: String,
    ariaLabel: String
  },
  created() {
    W(w), this.currentDir = this.$props.dir;
  },
  mounted() {
    this._progressStatus = f(this, "progressStatus"), this._progressStatusWrap = f(this, "progressStatusWrap"), this.currentDir = this.$props.dir !== void 0 ? this.$props.dir : O(this.$el) ? "rtl" : "ltr", this.animationFrame = g({
      duration: this.animationDuration,
      onStart: this.handleStart,
      onUpdate: this.handleUpdate,
      onEnd: this.handleEnd
    });
  },
  unmounted() {
    x(this.animationFrame);
  },
  data() {
    return {
      currentDir: void 0
    };
  },
  watch: {
    value: function(e, t) {
      this.prevValue = t, this.animationFrame = g({
        duration: this.animationDuration,
        onStart: this.handleStart,
        onUpdate: this.handleUpdate,
        onEnd: this.handleEnd
      });
    }
  },
  computed: {
    wrapperClass() {
      const {
        disabled: e,
        reverse: t,
        orientation: s,
        value: i
      } = this.$props, a = s === "vertical";
      return {
        "k-progressbar": !0,
        "k-progressbar-horizontal": !a,
        "k-progressbar-vertical": a,
        "k-progressbar-reverse": t,
        "k-progressbar-indeterminate": i === null,
        "k-disabled": e
      };
    },
    isVertical() {
      return this.orientation === "vertical";
    },
    animationDuration() {
      const {
        animation: e
      } = this.$props;
      return typeof e != "boolean" && e !== void 0 ? e.duration : e ? B : U;
    }
  },
  render() {
    const {
      disabled: e,
      labelVisible: t,
      labelPlacement: s,
      max: i,
      min: a,
      tabIndex: p,
      emptyStyle: S,
      emptyClassName: d,
      progressStyle: v,
      progressClassName: u
    } = this.$props, n = this.$props.value || 0, y = this.$props.value === null, N = A(n), k = {
      value: n
    }, $ = this.$props.labelRender ? R.call(this, this.$props.labelRender, D.call(this)) : void 0, _ = m("span", {
      class: "k-progress-status"
    }, N), V = C.call(this, {
      h: m,
      template: $,
      defaultRendering: _,
      additionalProps: k
    }), h = t ? this.$props.label ? r("span", {
      class: "k-progress-status"
    }, [this.$props.label]) : V : void 0, c = E("k-progress-status-wrap", {
      "k-progress-start": s === "start",
      "k-progress-center": s === "center",
      "k-progress-end": s === "end" || s === void 0
    });
    return r("div", {
      class: this.wrapperClass,
      dir: this.currentDir,
      tabindex: I(p, e),
      role: "progressbar",
      "aria-valuemin": a,
      "aria-valuemax": i,
      "aria-valuenow": y ? void 0 : n,
      "aria-disabled": e,
      "aria-label": this.$props.ariaLabel
    }, [r("span", {
      class: c + (d ? " " + d : ""),
      style: S
    }, [h]), r("div", {
      class: "k-progressbar-value k-selected",
      style: v,
      ref: b(this, "progressStatus")
    }, [r("span", {
      class: c + (u ? " " + u : ""),
      ref: b(this, "progressStatusWrap")
    }, [h])])]);
  },
  methods: {
    focus() {
      this.$el && this.$el.focus();
    },
    progressStatusElement() {
      return this._progressStatus;
    },
    progressStatusWrapElement() {
      return this._progressStatusWrap;
    },
    handleStart() {
      const e = o(this.min, this.max, this.prevValue);
      l(this._progressStatus, this._progressStatusWrap, e, this.isVertical);
    },
    handleUpdate(e) {
      const t = o(this.min, this.max, this.prevValue + (this.value - this.prevValue) * e);
      l(this._progressStatus, this._progressStatusWrap, t, this.isVertical);
    },
    handleEnd() {
      const e = o(this.min, this.max, this.value);
      l(this._progressStatus, this._progressStatusWrap, e, this.isVertical);
    }
  }
});
export {
  K as ProgressBar
};
