// @progress/kendo-vue-data-tools/navigation/constants.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2025 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
const d = "data-keyboardnavlevel", e = "data-keyboardnavscope", o = "data-keyboardnavheader", n = "data-keyboardnavbody", A = "data-keyboardnavid", b = "data-keyboardnavzone", r = "_filter", t = ["input:not([disabled]):not([type=hidden])", "select:not([disabled])", "textarea:not([disabled])", "button:not([disabled])", "a[href]", "area[href]", "summary", "iframe", "object", "embed", "audio[controls]", "video[controls]", "[contenteditable]"], E = [...t, "[tabindex]"], _ = [...t, "[tabindex]"].map((a) => a + ':not([tabindex="-1"])'), s = {
  [e]: !0
}, i = {
  [o]: !0
}, c = {
  [n]: !0
};
export {
  E as FOCUSABLE_ELEMENTS,
  n as KEYBOARD_NAV_DATA_BODY,
  o as KEYBOARD_NAV_DATA_HEADER,
  A as KEYBOARD_NAV_DATA_ID,
  d as KEYBOARD_NAV_DATA_LEVEL,
  e as KEYBOARD_NAV_DATA_SCOPE,
  b as KEYBOARD_NAV_DATA_ZONE,
  r as KEYBOARD_NAV_FILTER_COL_SUFFIX,
  _ as TABBABLE_ELEMENTS,
  c as tableKeyboardNavigationBodyAttributes,
  i as tableKeyboardNavigationHeaderAttributes,
  s as tableKeyboardNavigationScopeAttributes
};
