// @progress/kendo-vue-dateinputs/calendar/components/CalendarHeaderTitle.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2025 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
import { defineComponent as e, createVNode as o, isVNode as i } from "vue";
import { getDefaultSlots as r } from "@progress/kendo-vue-common";
import { Button as n } from "@progress/kendo-vue-buttons";
function l(t) {
  return typeof t == "function" || Object.prototype.toString.call(t) === "[object Object]" && !i(t);
}
const m = /* @__PURE__ */ e({
  name: "KendoHeaderTitle",
  emits: {
    click: (t) => !0
  },
  props: {
    id: String,
    value: String,
    view: Number
  },
  methods: {
    handleClick(t) {
      this.$emit("click", t);
    }
  },
  /**
   * @return
   * Returns a `<button />` element with the [`value`]({% slug api_dateinputs_calendarheadertitleprops %}#toc-value) of the title as a child.
   */
  render() {
    const t = r(this);
    return o(n, {
      role: "link",
      type: "button",
      id: this.$props.id,
      fillMode: "flat",
      themeColor: "primary",
      onClick: this.handleClick,
      tabIndex: 0
    }, l(t) ? t : {
      default: () => [t]
    });
  }
});
export {
  m as CalendarHeaderTitle
};
