// @progress/kendo-vue-dateinputs/calendar/components/CalendarWeekCell.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2025 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
import { defineComponent as i, createVNode as o } from "vue";
import { getDefaultSlots as r } from "@progress/kendo-vue-common";
const n = /* @__PURE__ */ i({
  props: {
    id: String,
    value: Number,
    className: String
  },
  emits: {
    click: (e) => !0
  },
  /**
   * @return
   * Returns a `<td />` element with the [`value`]({% slug api_dateinputs_calendarweekcellprops %}#toc-value) as a child.
   */
  render() {
    const e = r(this);
    return o("td", {
      onClick: this.handleClick,
      id: this.$props.id
    }, [this.$props.value !== void 0 ? this.$props.value : e]);
  },
  methods: {
    handleClick(e) {
      const {
        value: t
      } = this.$props;
      this.$emit("click", t, e);
    }
  }
});
export {
  n as CalendarWeekCell
};
