// @progress/kendo-vue-layout/card/Card.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2025 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
import { defineComponent as a, createVNode as o } from "vue";
import { cardOrientation as t, cardType as e } from "./interfaces/Enums.mjs";
import { packageMetadata as i } from "../package-metadata.mjs";
import { getDefaultSlots as n, validatePackage as p } from "@progress/kendo-vue-common";
const u = /* @__PURE__ */ a({
  name: "KendoCard",
  props: {
    dir: String,
    type: {
      type: String,
      default: e.DEFAULT,
      validator: function(r) {
        return ["default", "primary", "info", "success", "warning", "error"].includes(r);
      }
    },
    orientation: {
      type: String,
      default: t.VERTICAL,
      validator: function(r) {
        return ["horizontal", "vertical"].includes(r);
      }
    }
  },
  created() {
    p(i);
  },
  computed: {
    wrapperClass() {
      const r = this.$props.orientation !== t.HORIZONTAL ? "vertical" : "horizontal";
      return {
        "k-card": !0,
        [`k-card-${this.$props.type}`]: this.$props.type !== e.DEFAULT,
        [`k-card-${r}`]: !0
      };
    }
  },
  render() {
    const r = n(this);
    return o("div", {
      dir: this.$props.dir,
      class: this.wrapperClass
    }, [r]);
  }
});
export {
  u as Card
};
