// @progress/kendo-vue-layout/card/CardActions.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2025 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
import { defineComponent as o, createVNode as e } from "vue";
import { getDefaultSlots as n } from "@progress/kendo-vue-common";
import { cardOrientation as r, cardActionsLayout as a } from "./interfaces/Enums.mjs";
const d = /* @__PURE__ */ o({
  name: "KendoCardActions",
  props: {
    layout: {
      type: String,
      default: a.START,
      validator: function(t) {
        return ["stretched", "start", "center", "end"].includes(t);
      }
    },
    orientation: {
      type: String,
      default: r.HORIZONTAL,
      validator: function(t) {
        return ["horizontal", "vertical"].includes(t);
      }
    }
  },
  computed: {
    wrapperClass() {
      return {
        "k-card-actions": !0,
        "k-actions": !0,
        [`k-actions-${this.$props.layout}`]: !0,
        [`k-actions-${this.$props.orientation !== r.HORIZONTAL ? "vertical" : "horizontal"}`]: !0
      };
    }
  },
  render() {
    const t = n(this);
    return e("div", {
      class: this.wrapperClass
    }, [t]);
  }
});
export {
  d as CardActions
};
