// @progress/kendo-vue-dateinputs/messages/main.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2025 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
const e = "calendar.prevView", t = "calendar.nextView", a = "dateinput.increment", c = "dateinput.decrement", i = "calendar.today", n = "datepicker.toggleCalendar", o = "daterangepicker.swapStartEnd", r = "daterangepicker.start", s = "daterangepicker.end", l = "daterangepicker.separator", d = "datetimepicker.toggleDateTimeSelector", g = "timepicker.now", m = "timepicker.selectNow", p = "timepicker.cancel", k = "timepicker.set", w = "timepicker.toggleTimeSelector", S = "timepicker.toggleClock", T = "datetimepicker.date", v = "datetimepicker.time", C = "datetimepicker.cancel", u = "datetimepicker.set", N = {
  [i]: "Today",
  [g]: "NOW",
  [k]: "Set",
  [p]: "Cancel",
  [T]: "Date",
  [v]: "Time",
  [C]: "Cancel",
  [u]: "Set",
  [r]: "Start",
  [s]: "End",
  [l]: " ",
  [m]: "Select Now",
  [w]: "Toggle TimeSelector",
  [S]: "Toggle Clock",
  [a]: "Increase value",
  [c]: "Decrease value",
  [n]: "Toggle calendar",
  [e]: "Navigate to previous view",
  [t]: "Navigate to next view",
  [o]: "Swap start and end values",
  [d]: "Toggle date-time selector"
};
export {
  T as date,
  C as dateTimePickerCancel,
  u as dateTimePickerSet,
  c as decreaseValue,
  s as end,
  a as increaseValue,
  N as messages,
  t as nextView,
  g as now,
  e as prevView,
  m as selectNow,
  l as separator,
  r as start,
  o as swapStartEnd,
  v as time,
  p as timePickerCancel,
  k as timePickerSet,
  i as today,
  n as toggleCalendar,
  S as toggleClock,
  d as toggleDateTimeSelector,
  w as toggleTimeSelector
};
