// @progress/kendo-vue-data-tools/filter/filters/BooleanFilter.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2025 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
import { defineComponent as t, createVNode as r } from "vue";
import { EnumFilter as a } from "./EnumFilter.mjs";
import { filterIsTrue as n, filterIsFalse as i } from "../../messages/main.mjs";
const s = /* @__PURE__ */ t({
  name: "KendoBooleanFilter",
  props: {
    filter: {
      type: Object,
      required: !0
    },
    dataItems: {
      type: Array,
      default: function() {
      }
    },
    defaultItem: Object,
    onFilterchange: Function
  },
  data() {
    return {
      currentData: [{
        text: n,
        value: !0
      }, {
        text: i,
        value: !1
      }]
    };
  },
  render() {
    return r(a, {
      filter: this.$props.filter,
      onFilterchange: this.handleFilterChange,
      dataItems: this.$props.dataItems || this.currentData,
      defaultItem: this.$props.defaultItem
    }, null);
  },
  methods: {
    handleFilterChange(e) {
      this.$emit("filterchange", e);
    }
  }
});
export {
  s as BooleanFilter
};
