// @progress/kendo-vue-data-tools/filter/filters/DateFilter.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2025 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
import { defineComponent as t, createVNode as r } from "vue";
import { DatePicker as n } from "@progress/kendo-vue-dateinputs";
const l = /* @__PURE__ */ t({
  name: "KendoDateFilter",
  props: {
    filter: {
      type: Object,
      required: !0
    },
    onFilterchange: Function
  },
  render() {
    const e = this.$props.filter;
    return r(n, {
      value: e.value || null,
      onChange: this.onChange
    }, null);
  },
  methods: {
    onChange(e) {
      this.$emit("filterchange", {
        nextFilter: {
          ...this.$props.filter,
          value: e.value
        }
      });
    }
  }
});
export {
  l as DateFilter
};
