// @progress/kendo-vue-data-tools/filter/filters/NumericFilter.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2025 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
import { defineComponent as i, createVNode as o } from "vue";
import { NumericTextBox as n } from "@progress/kendo-vue-inputs";
import { provideLocalizationService as l } from "@progress/kendo-vue-intl";
import { filterNumericFilterAriaLabel as r, messages as a } from "../../messages/main.mjs";
const f = /* @__PURE__ */ i({
  name: "kendoNumericFilter",
  props: {
    filter: {
      type: Object,
      required: !0
    },
    onFilterchange: Function
  },
  inject: {
    kendoLocalizationService: {
      default: null
    }
  },
  render() {
    const e = this.$props.filter, t = l(this).toLanguageString(r, a[r]);
    return o(n, {
      value: typeof e.value == "number" ? e.value : null,
      onChange: this.onChange,
      ariaLabel: t
    }, null);
  },
  methods: {
    onChange(e) {
      this.$emit("filterchange", {
        nextFilter: {
          ...this.$props.filter,
          value: e.value
        }
      });
    }
  }
});
export {
  f as NumericFilter
};
