// @progress/kendo-vue-data-tools/filter/filters/TextFilter.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2025 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
import { defineComponent as i, createVNode as o } from "vue";
import { TextBox as n } from "@progress/kendo-vue-inputs";
import { provideLocalizationService as a } from "@progress/kendo-vue-intl";
import { filterTextFilterAriaLabel as t, messages as l } from "../../messages/main.mjs";
const u = /* @__PURE__ */ i({
  name: "kendoTextFilter",
  props: {
    filter: {
      type: Object,
      required: !0
    },
    ariaLabel: String,
    onFilterchange: Function
  },
  inject: {
    kendoLocalizationService: {
      default: null
    }
  },
  render() {
    const e = a(this), {
      ariaLabel: r = e.toLanguageString(t, l[t])
    } = this.$props;
    return o(n, {
      value: this.$props.filter.value || "",
      onInput: this.onChange,
      ariaLabel: r
    }, null);
  },
  methods: {
    onChange(e) {
      this.$emit("filterchange", {
        nextFilter: {
          ...this.$props.filter,
          value: e.value
        }
      });
    }
  }
});
export {
  u as TextFilter
};
