// @progress/kendo-vue-grid/messages/main.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2025 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
const a = "grid.noRecords", i = "grid.selectAllRows", o = "grid.pagerInfo", l = "grid.pagerFirstPage", s = "grid.pagerPreviousPage", g = "grid.pagerNextPage", n = "grid.pagerLastPage", d = "grid.pagerItemsPerPage", p = "grid.pagerPageSelection", c = "grid.pagerPage", u = "grid.pagerOf", f = "grid.pagerTotalPages", A = "grid.groupPanelEmpty", b = "grid.columnMenu", m = "grid.filterClearButton", L = "grid.filterSubmitButton", C = "grid.filterTitle", I = "grid.sortAscending", h = "grid.sortDescending", O = "grid.searchPlaceholder", P = "grid.filterCheckAll", F = "grid.filterChooseOperator", S = "grid.filterSelectedItems", D = "grid.sortAriaLabel", E = "grid.sortableColumnAriaLabel", x = "grid.sortableColumnAscendingAriaLabel", q = "grid.sortableColumnDescendingAriaLabel", N = "grid.filterAriaLabel", G = "grid.numericFilterAriaLabel", y = "grid.dateFilterAriaLabel", B = "grid.textFilterAriaLabel", T = "grid.booleanFilterAriaLabel", M = "grid.groupHeaderAriaLabel", w = "grid.groupCaretAriaLabelCollapse", R = "grid.groupCaretAriaLabelExpand", v = "grid.expandDetailTemplateAriaLabel", k = "grid.collapseDetailTemplateAriaLabel", H = {
  [i]: "Select All Rows",
  [a]: "No records available",
  [A]: "Drag a column header and drop it here to group by that column",
  [b]: "Column Menu",
  [m]: "Clear",
  [L]: "Filter",
  [C]: "Filter",
  [I]: "Sort Ascending",
  [h]: "Sort Descending",
  [d]: "items per page",
  [p]: "Select page",
  [o]: "{0} - {1} of {2} items",
  [l]: "Go to the first page",
  [s]: "Go to the previous page",
  [g]: "Go to the next page",
  [n]: "Go to the last page",
  [c]: "Page",
  [u]: "of",
  [f]: "{0}",
  [O]: "Search",
  [P]: "Check All",
  [F]: "Choose Operator",
  [S]: "selected items",
  [D]: "Sortable",
  [E]: "Sortable Column",
  [x]: "Sorted in ascending order",
  [q]: "Sorted in descending order",
  [N]: "Filter",
  [G]: "Numeric Filter",
  [y]: "Date Filter",
  [B]: "Text Filter",
  [T]: "Boolean Filter",
  [M]: "Group Header",
  [w]: "Collapse Group",
  [R]: "Expand Group",
  [v]: "Expand Details",
  [k]: "Collapse Details",
  "grid.filterEqOperator": "Is equal to",
  "grid.filterNotEqOperator": "Is not equal to",
  "grid.filterIsNullOperator": "Is null",
  "grid.filterIsNotNullOperator": "Is not null",
  "grid.filterIsEmptyOperator": "Is empty",
  "grid.filterIsNotEmptyOperator": "Is not empty",
  "grid.filterStartsWithOperator": "Starts with",
  "grid.filterContainsOperator": "Contains",
  "grid.filterNotContainsOperator": "Does not contain",
  "grid.filterEndsWithOperator": "Ends with",
  "grid.filterGteOperator": "Is greater than or equal to",
  "grid.filterGtOperator": "Is greater than",
  "grid.filterLteOperator": "Is less than or equal to",
  "grid.filterLtOperator": "Is less than",
  "grid.filterIsTrue": "Is true",
  "grid.filterIsFalse": "Is false",
  "grid.filterBooleanAll": "(All)",
  "grid.filterAfterOrEqualOperator": "Is after or equal to",
  "grid.filterAfterOperator": "Is after",
  "grid.filterBeforeOperator": "Is before",
  "grid.filterBeforeOrEqualOperator": "Is before or equal to",
  "grid.filterAndLogic": "And",
  "grid.filterOrLogic": "Or"
};
function z(r) {
  const e = r.replace(
    /^pager\.([a-z])/,
    (W, t) => "grid.pager" + t.toUpperCase()
  );
  return { messageKey: e, defaultMessage: H[e] };
}
export {
  T as booleanFilterAriaLabel,
  k as collapseDetailAriaLabel,
  b as columnMenu,
  y as dateFilterAriaLabel,
  v as expandDetailAriaLabel,
  N as filterAriaLabel,
  P as filterCheckAll,
  F as filterChooseOperator,
  m as filterClearButton,
  S as filterSelectedItems,
  L as filterSubmitButton,
  C as filterTitle,
  w as groupCaretAriaLabelCollapse,
  R as groupCaretAriaLabelExpand,
  M as groupHeaderAriaLabel,
  A as groupPanelEmpty,
  H as messages,
  a as noRecords,
  G as numericFilterAriaLabel,
  l as pagerFirstPage,
  o as pagerInfo,
  d as pagerItemPerPage,
  n as pagerLastPage,
  z as pagerMessagesMap,
  g as pagerNextPage,
  u as pagerOf,
  c as pagerPage,
  p as pagerPageSelection,
  s as pagerPreviousPage,
  f as pagerTotalPages,
  O as searchPlaceholder,
  i as selectAllRows,
  D as sortAriaLabel,
  I as sortAscending,
  h as sortDescending,
  E as sortableColumnAriaLabel,
  x as sortableColumnAscendingAriaLabel,
  q as sortableColumnDescendingAriaLabel,
  B as textFilterAriaLabel
};
