// @progress/kendo-vue-treeview/TreeViewDragClue.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2025 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
import { defineComponent as n, createVNode as o } from "vue";
import { Icon as a, getIconName as l } from "@progress/kendo-vue-common";
import { cancelIcon as r } from "@progress/kendo-svg-icons";
const h = /* @__PURE__ */ n({
  name: "KendoTreeViewDragClue",
  data() {
    return {
      visible: !1,
      top: 0,
      left: 0,
      text: "",
      operationIconName: "cancel",
      operationSvg: r
    };
  },
  render() {
    const e = {
      display: "block",
      position: "absolute",
      zIndex: 2e4,
      padding: "4px 6px"
    }, t = {
      top: this.top + "px",
      left: this.left + "px"
    };
    return this.visible && o("div", {
      class: "k-header k-drag-clue",
      style: {
        ...e,
        ...t
      }
    }, [o(a, {
      name: l(this.operationIconName),
      icon: this.operationSvg,
      class: "k-drag-status"
    }, null), this.text]);
  },
  methods: {
    show(e, t, s, i) {
      this.visible = !0, this.top = e, this.left = t, this.text = s, typeof i == "string" ? this.operationIconName = i : this.operationSvg = i;
    },
    hide() {
      this.visible = !1;
    }
  }
});
export {
  h as TreeViewDragClue
};
