// @progress/kendo-vue-layout/gridlayout/GridLayoutItem.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2025 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
import { defineComponent as a, h as n, createVNode as i } from "vue";
import { packageMetadata as d } from "../package-metadata.mjs";
import { getTemplate as m, validatePackage as p, guid as c } from "@progress/kendo-vue-common";
const g = /* @__PURE__ */ a({
  name: "KendoGridLayoutItem",
  props: {
    content: [String, Function, Object],
    id: String,
    col: Number,
    colSpan: Number,
    row: Number,
    rowSpan: Number
  },
  created() {
    p(d), this.itemId = c();
  },
  computed: {
    gridLayoutItemStyles() {
      const {
        row: e,
        col: o,
        rowSpan: t,
        colSpan: r
      } = this.$props;
      return {
        gridArea: `${e || "auto"} / ${o || "auto"} / ${t ? "span " + t : "auto"} / ${r ? "span " + r : "auto"}`
      };
    }
  },
  render() {
    const {
      id: e,
      content: o
    } = this.$props, t = m.call(this, {
      h: n,
      template: o
    });
    return i("div", {
      style: this.gridLayoutItemStyles,
      id: e || this.itemId
    }, [t]);
  }
});
export {
  g as GridLayoutItem
};
