// @progress/kendo-vue-common/index.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2025 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
import { classNames as t } from "./classNames.mjs";
import { guid as a } from "./guid.mjs";
import { Keys as p } from "./keys.mjs";
import { canUseDOM as l } from "./canUseDOM.mjs";
import { getRef as s, setRef as i } from "./refs.mjs";
import { firstFocusableChild as c, focusContainer as g, focusFirstFocusableChild as d, focusLastFocusableChild as I, lastFocusableChild as E } from "./focus.mjs";
import { Draggable as h } from "./Draggable.mjs";
import { clone as D, cloneArray as S, cloneDate as b, cloneObject as v, cloneValue as T } from "./clone.mjs";
import { isObject as F } from "./isObject.mjs";
import { noop as R } from "./noop.mjs";
import { getter as N } from "./getter.mjs";
import { setter as P } from "./setter.mjs";
import { getListeners as V, hasListener as k } from "./listeners.mjs";
import { getTemplate as U, templateDefinition as y, templateRendering as K } from "./templateRendering.mjs";
import { getTabIndex as j } from "./getTabIndex.mjs";
import { getDefaultSlots as G } from "./defaultSlots.mjs";
import { getDir as q, isRtl as z } from "./isRtl.mjs";
import { getLicenseMessage as J, shouldShowValidationUI as Q, validatePackage as Y } from "./validate-package.mjs";
import { AdaptiveModeProvider as $ } from "./providers/AdaptiveModeProvider.mjs";
import { ADAPTIVE_MEDIUM_BREAKPOINT as oe, ADAPTIVE_SMALL_BREAKPOINT as te, FIELD_REGEX as re } from "./constants/main.mjs";
import { kendoThemeMaps as me } from "./theme.mjs";
import { extendDataItem as fe, getItemPath as le, getNestedValue as xe, mapTree as se, mapTreeItem as ie } from "./treeDataOperations.mjs";
import { BrowserSupportService as ce } from "./browser-support.service.mjs";
import { getScrollbarWidth as de, setScrollbarWidth as Ie } from "./scrollbarWidth.mjs";
import { hasRelativeStackingContext as ue } from "./hasRelativeStackingContext.mjs";
import { Icon as Ae } from "./icons/Icon.mjs";
import { FontIcon as Se } from "./icons/FontIcon.mjs";
import { SvgIcon as ve } from "./icons/SvgIcon.mjs";
import { getIconName as Le } from "./icons/getIconName.mjs";
import { FOCUSABLE_ELEMENTS as Me, Navigation as Re } from "./navigation.mjs";
import { getActiveElement as Ne, getInnerActiveElement as Oe } from "./getActiveElement.mjs";
import { WatermarkOverlay as _e } from "./watermark/WatermarkOverlay.mjs";
export {
  oe as ADAPTIVE_MEDIUM_BREAKPOINT,
  te as ADAPTIVE_SMALL_BREAKPOINT,
  $ as AdaptiveModeProvider,
  ce as BrowserSupportService,
  h as Draggable,
  re as FIELD_REGEX,
  Me as FOCUSABLE_ELEMENTS,
  Se as FontIcon,
  Ae as Icon,
  p as Keys,
  Re as Navigation,
  ve as SvgIcon,
  _e as WatermarkOverlay,
  l as canUseDOM,
  t as classNames,
  D as clone,
  S as cloneArray,
  b as cloneDate,
  v as cloneObject,
  T as cloneValue,
  fe as extendDataItem,
  c as firstFocusableChild,
  g as focusContainer,
  d as focusFirstFocusableChild,
  I as focusLastFocusableChild,
  Ne as getActiveElement,
  G as getDefaultSlots,
  q as getDir,
  Le as getIconName,
  Oe as getInnerActiveElement,
  le as getItemPath,
  J as getLicenseMessage,
  V as getListeners,
  xe as getNestedValue,
  s as getRef,
  de as getScrollbarWidth,
  j as getTabIndex,
  U as getTemplate,
  N as getter,
  a as guid,
  k as hasListener,
  ue as hasRelativeStackingContext,
  F as isObject,
  z as isRtl,
  me as kendoThemeMaps,
  E as lastFocusableChild,
  se as mapTree,
  ie as mapTreeItem,
  R as noop,
  i as setRef,
  Ie as setScrollbarWidth,
  P as setter,
  Q as shouldShowValidationUI,
  y as templateDefinition,
  K as templateRendering,
  Y as validatePackage
};
