// @progress/kendo-vue-layout/splitter/SplitterPane.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2025 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
import { defineComponent as f, h as m, createVNode as t } from "vue";
import { classNames as S, getTemplate as g } from "@progress/kendo-vue-common";
const b = /* @__PURE__ */ f({
  name: "KendoSplitterPane",
  props: {
    orientation: {
      type: String,
      default: function() {
      },
      validator: function(e) {
        return ["vertical", "horizontal"].includes(e);
      }
    },
    overlay: Boolean,
    containsSplitter: Boolean,
    size: String,
    min: String,
    max: String,
    resizable: Boolean,
    collapsible: Boolean,
    collapsed: Boolean,
    scrollable: Boolean,
    keepMounted: Boolean,
    content: [Object, Function, String]
  },
  render() {
    const {
      size: e,
      collapsed: n,
      overlay: l,
      containsSplitter: o,
      collapsible: a,
      resizable: i,
      scrollable: r,
      keepMounted: s
    } = this.$props, c = e && e.length > 0, p = {
      flexBasis: e
    }, d = S("k-pane", {
      "k-hidden": n,
      hidden: n,
      "k-pane-flex": o,
      "k-pane-static": !i && !a || c,
      "k-scrollable": r
    }), u = g.call(this, {
      h: m,
      template: this.$props.content,
      defaultRendering: null
    });
    return t("div", {
      style: p,
      class: d,
      role: "group"
    }, [!n || s ? u : void 0, l ? t("div", {
      class: "k-splitter-overlay k-overlay"
    }, null) : void 0]);
  }
});
export {
  b as SplitterPane
};
