// @progress/kendo-vue-data-tools/messages/main.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2025 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
const qe = "grid.columnMenu", e = "pager.info", r = "pager.firstPage", t = "pager.previousPage", o = "pager.nextPage", l = "pager.lastPage", n = "pager.itemsPerPage", a = "pager.pageInputAriaLabel", i = "pager.smallPageSelectorAriaLabel", s = "pager.pageSelection", u = "pager.page", p = "pager.of", c = "pager.totalPages", f = "filter.eqOperator", O = "filter.groupFilterAriaLabel", g = "filter.notEqOperator", m = "filter.isNullOperator", d = "filter.isNotNullOperator", I = "filter.isEmptyOperator", A = "filter.isNotEmptyOperator", M = "filter.startsWithOperator", F = "filter.containsOperator", L = "filter.notContainsOperator", E = "filter.endsWithOperator", h = "filter.gteOperator", b = "filter.gtOperator", q = "filter.lteOperator", N = "filter.ltOperator", P = "filter.isTrue", S = "filter.isFalse", C = "filter.afterOrEqualOperator", w = "filter.afterOperator", G = "filter.beforeOperator", x = "filter.beforeOrEqualOperator", y = "filter.andLogic", D = "filter.orLogic", B = "filter.addExpression", T = "filter.addGroup", W = "filter.close", v = "columnMenu.filterChooseOperator", R = "columnMenu.filterClearButton", k = "columnMenu.filterSubmitButton", j = "columnMenu.filterTitle", z = "columnMenu.sortAscending", H = "columnMenu.sortDescending", J = "columnMenu.filterEqOperator", K = "columnMenu.filterNotEqOperator", Q = "columnMenu.filterIsNullOperator", U = "columnMenu.filterIsNotNullOperator", V = "columnMenu.filterIsEmptyOperator", X = "columnMenu.filterIsNotEmptyOperator", Y = "columnMenu.filterStartsWithOperator", Z = "columnMenu.filterContainsOperator", _ = "columnMenu.filterNotContainsOperator", $ = "columnMenu.filterEndsWithOperator", ee = "columnMenu.filterGteOperator", re = "columnMenu.filterGtOperator", te = "columnMenu.filterLteOperator", oe = "columnMenu.filterLtOperator", le = "columnMenu.filterIsTrue", ne = "columnMenu.filterAfterOrEqualOperator", ae = "columnMenu.filterAfterOperator", ie = "columnMenu.filterBeforeOperator", se = "columnMenu.filterBeforeOrEqualOperator", ue = "columnMenu.filterAndLogic", pe = "columnMenu.filterOrLogic", ce = "grid.searchPlaceholder", fe = "grid.selectAllRows", Oe = "grid.filterCheckAll", ge = "sort.ariaLabel", me = "filter.ariaLabel", de = "filter.groupAriaLabel", Ie = "filter.expressionAriaLabel", Ae = "filter.expressionDropdownAriaLabel", Me = "filter.expressionOperatorDropdownAriaLabel", Fe = "filter.enumFilterDropdownAriaLabel", Le = "filter.numericFilterAriaLabel", Ee = "filter.textFilterAriaLabel", he = "filter.dateFilterAriaLabel", be = "filter.dropdownOperatorAriaLabel", Ne = {
  [O]: "Group Filter",
  [i]: "Select page",
  [a]: "Pager - Page number input",
  [n]: "items per page",
  [s]: "Select page",
  [e]: "{0} - {1} of {2} items",
  [r]: "Go to the first page",
  [t]: "Go to the previous page",
  [o]: "Go to the next page",
  [l]: "Go to the last page",
  [u]: "Page",
  [p]: "of",
  [c]: "{0}",
  [ce]: "Search",
  [fe]: "Select All Rows",
  [Oe]: "Check All",
  [ge]: "Sortable",
  [me]: "Filter",
  [f]: "Is equal to",
  [g]: "Is not equal to",
  [m]: "Is null",
  [d]: "Is not null",
  [I]: "Is empty",
  [A]: "Is not empty",
  [M]: "Starts with",
  [F]: "Contains",
  [L]: "Does not contain",
  [E]: "Ends with",
  [h]: "Is greater than or equal to",
  [b]: "Is greater than",
  [q]: "Is less than or equal to",
  [N]: "Is less than",
  [P]: "Is true",
  [S]: "Is false",
  [C]: "Is after or equal to",
  [w]: "Is after",
  [G]: "Is before",
  [x]: "Is before or equal to",
  [y]: "And",
  [D]: "Or",
  [B]: "Add Expression",
  [T]: "Add Group",
  [W]: "Remove",
  [de]: "Filter toolbar",
  [Ie]: "Filter expression row",
  [Ae]: "Filter dropdown",
  [Me]: "Filter operator dropdown",
  [Fe]: "Select True/False",
  [Le]: "Enter number",
  [Ee]: "Enter text",
  [he]: "Enter date",
  [be]: "Choose Operator",
  [v]: "Choose Operator",
  [R]: "Clear",
  [k]: "Filter",
  [j]: "Filter",
  [z]: "Sort Ascending",
  [H]: "Sort Descending",
  [J]: "Is equal to",
  [K]: "Is not equal to",
  [Q]: "Is null",
  [U]: "Is not null",
  [V]: "Is empty",
  [X]: "Is not empty",
  [Y]: "Starts with",
  [Z]: "Contains",
  [_]: "Does not contain",
  [$]: "Ends with",
  [ee]: "Is greater than or equal to",
  [re]: "Is greater than",
  [te]: "Is less than or equal to",
  [oe]: "Is less than",
  [le]: "Is true",
  [ne]: "Is after or equal to",
  [ae]: "Is after",
  [ie]: "Is before",
  [se]: "Is before or equal to",
  [ue]: "And",
  [pe]: "Or",
  "grid.filterEqOperator": "Is equal to",
  "grid.filterNotEqOperator": "Is not equal to",
  "grid.filterIsNullOperator": "Is null",
  "grid.filterIsNotNullOperator": "Is not null",
  "grid.filterIsEmptyOperator": "Is empty",
  "grid.filterIsNotEmptyOperator": "Is not empty",
  "grid.filterStartsWithOperator": "Starts with",
  "grid.filterContainsOperator": "Contains",
  "grid.filterNotContainsOperator": "Does not contain",
  "grid.filterEndsWithOperator": "Ends with",
  "grid.filterGteOperator": "Is greater than or equal to",
  "grid.filterGtOperator": "Is greater than",
  "grid.filterLteOperator": "Is less than or equal to",
  "grid.filterLtOperator": "Is less than",
  "grid.filterIsTrue": "Is true",
  "grid.filterIsFalse": "Is false",
  "grid.filterBooleanAll": "(All)",
  "grid.filterAfterOrEqualOperator": "Is after or equal to",
  "grid.filterAfterOperator": "Is after",
  "grid.filterBeforeOperator": "Is before",
  "grid.filterBeforeOrEqualOperator": "Is before or equal to",
  "grid.filterAndLogic": "And",
  "grid.filterOrLogic": "Or"
};
export {
  qe as columnMenu,
  ae as columnMenuFilterAfterOperator,
  ne as columnMenuFilterAfterOrEqualOperator,
  ue as columnMenuFilterAndLogic,
  ie as columnMenuFilterBeforeOperator,
  se as columnMenuFilterBeforeOrEqualOperator,
  v as columnMenuFilterChooseOperator,
  R as columnMenuFilterClearButton,
  Z as columnMenuFilterContainsOperator,
  $ as columnMenuFilterEndsWithOperator,
  J as columnMenuFilterEqOperator,
  re as columnMenuFilterGtOperator,
  ee as columnMenuFilterGteOperator,
  V as columnMenuFilterIsEmptyOperator,
  X as columnMenuFilterIsNotEmptyOperator,
  U as columnMenuFilterIsNotNullOperator,
  Q as columnMenuFilterIsNullOperator,
  le as columnMenuFilterIsTrue,
  oe as columnMenuFilterLtOperator,
  te as columnMenuFilterLteOperator,
  _ as columnMenuFilterNotContainsOperator,
  K as columnMenuFilterNotEqOperator,
  pe as columnMenuFilterOrLogic,
  Y as columnMenuFilterStartsWithOperator,
  k as columnMenuFilterSubmitButton,
  j as columnMenuFilterTitle,
  z as columnMenuSortAscending,
  H as columnMenuSortDescending,
  B as filterAddExpression,
  T as filterAddGroup,
  w as filterAfterOperator,
  C as filterAfterOrEqualOperator,
  y as filterAndLogic,
  me as filterAriaLabel,
  G as filterBeforeOperator,
  x as filterBeforeOrEqualOperator,
  Oe as filterCheckAll,
  W as filterClose,
  F as filterContainsOperator,
  he as filterDateFilterAriaLabel,
  be as filterDropDownOperatorAriaLabel,
  E as filterEndsWithOperator,
  Fe as filterEnumFilterDropdownAriaLabel,
  f as filterEqOperator,
  Ie as filterExpressionAriaLabel,
  Ae as filterExpressionDropdownAriaLabel,
  Me as filterExpressionOperatorDropdownAriaLabel,
  de as filterGroupAriaLabel,
  O as filterGroupFilterAriaLabel,
  b as filterGtOperator,
  h as filterGteOperator,
  I as filterIsEmptyOperator,
  S as filterIsFalse,
  A as filterIsNotEmptyOperator,
  d as filterIsNotNullOperator,
  m as filterIsNullOperator,
  P as filterIsTrue,
  N as filterLtOperator,
  q as filterLteOperator,
  L as filterNotContainsOperator,
  g as filterNotEqOperator,
  Le as filterNumericFilterAriaLabel,
  D as filterOrLogic,
  M as filterStartsWithOperator,
  Ee as filterTextFilterAriaLabel,
  Ne as messages,
  r as pagerFirstPage,
  e as pagerInfo,
  n as pagerItemPerPage,
  l as pagerLastPage,
  o as pagerNextPage,
  p as pagerOf,
  u as pagerPage,
  a as pagerPageInputAriaLabel,
  s as pagerPageSelection,
  t as pagerPreviousPage,
  i as pagerSmallPageSelectorAriaLabel,
  c as pagerTotalPages,
  ce as searchPlaceholder,
  fe as selectAllRows,
  ge as sortAriaLabel
};
