// @progress/kendo-vue-dropdowns/common/ClearButton.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2025 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
import { defineComponent as t, createVNode as o, inject as i } from "vue";
import { xIcon as r } from "@progress/kendo-svg-icons";
import { provideLocalizationService as c } from "@progress/kendo-vue-intl";
import { clear as n, messages as a } from "../messages/main.mjs";
import { Icon as l } from "@progress/kendo-vue-common";
const k = /* @__PURE__ */ t({
  name: "clear-button",
  inject: {
    kendoLocalizationService: {
      default: null
    }
  },
  props: {
    onClearclick: Function
  },
  setup() {
    return {
      kendoLocalizationService: i("kendoLocalizationService", {})
    };
  },
  methods: {
    onMouseDown(e) {
      e.preventDefault();
    },
    onClickHandler(e) {
      this.$emit("clearclick", e);
    }
  },
  render() {
    const e = c(this).toLanguageString(n, a[n]);
    return o("span", {
      class: "k-clear-value",
      role: "button",
      onClick: this.onClickHandler,
      onMousedown: this.onMouseDown,
      tabindex: -1,
      title: e,
      key: "clearbutton"
    }, [o(l, {
      name: "x",
      icon: r
    }, null)]);
  }
});
export {
  k as ClearButton
};
