// @progress/kendo-vue-dropdowns/common/GroupStickyHeader.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2025 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
import { defineComponent as i, createVNode as e, h as p } from "vue";
import { getTemplate as s } from "@progress/kendo-vue-common";
const n = /* @__PURE__ */ i({
  name: "KendoGroupStickyHeader",
  inheritAttrs: !1,
  props: {
    group: String,
    render: Object
  },
  render() {
    const {
      group: r,
      render: t
    } = this.$props, o = e("div", {
      class: "k-list-group-sticky-header"
    }, [e("div", {
      class: "k-list-header-text"
    }, [r])]);
    return s.call(this, {
      h: p,
      template: t,
      defaultRendering: o,
      additionalProps: {
        ...this.$props
      }
    });
  }
});
export {
  n as GroupStickyHeader
};
