// @progress/kendo-vue-dropdowns/common/ListDefaultItem.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2025 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
import { defineComponent as i, createVNode as n } from "vue";
import { classNames as l } from "@progress/kendo-vue-common";
import { getItemValue as s } from "./utils.mjs";
const a = /* @__PURE__ */ i({
  name: "list-default-item",
  props: {
    defaultItem: [Object, String],
    textField: String,
    selected: Boolean,
    onDefaultitemclick: Function
  },
  methods: {
    onMouseDown(e) {
      e.preventDefault();
    },
    onClick(e) {
      this.$emit("defaultitemclick", e);
    }
  },
  render() {
    const {
      selected: e,
      defaultItem: t,
      textField: o
    } = this.$props;
    return n("div", {
      onClick: this.onClick,
      onMousedown: this.onMouseDown,
      style: {
        position: "unset"
      },
      class: l("k-list-optionlabel", {
        "k-selected": e
      })
    }, [s(t, o) || ""]);
  }
});
export {
  a as ListDefaultItem
};
