// @progress/kendo-vue-dropdowns/common/ListGroupItem.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2025 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
import { defineComponent as s, createVNode as e, h as p } from "vue";
import { getTemplate as d } from "@progress/kendo-vue-common";
const m = /* @__PURE__ */ s({
  name: "KendoListItem",
  inheritAttrs: !1,
  props: {
    id: String,
    group: String,
    render: Object
  },
  render() {
    const {
      id: t,
      group: r,
      virtual: i,
      render: n
    } = this.$props, o = e("li", {
      id: t,
      role: "group",
      class: "k-list-group-item",
      style: {
        position: i ? "relative" : "unset"
      }
    }, [e("span", {
      class: n ? void 0 : "k-list-item-text"
    }, [r])]);
    return d.call(this, {
      h: p,
      template: this.$props.render,
      defaultRendering: o,
      additionalProps: {
        ...this.$props
      }
    });
  }
});
export {
  m as ListGroupItem
};
