// @progress/kendo-vue-dropdowns/common/ListContainer.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2025 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
import { defineComponent as S, createVNode as t } from "vue";
import { Popup as b } from "@progress/kendo-vue-popup";
import { getDefaultSlots as v } from "@progress/kendo-vue-common";
const O = /* @__PURE__ */ S({
  name: "list-container",
  emits: {
    mousedown: (e) => !0,
    blur: (e) => !0,
    open: (e) => !0,
    close: (e) => !0
  },
  props: {
    width: [String, Number],
    dir: String,
    itemsCount: Number,
    popupSettings: {
      type: Object,
      default: function() {
        return {
          animate: !0,
          height: "200px"
        };
      }
    }
  },
  created() {
    this.kendoAnchorRef = void 0;
  },
  methods: {
    onMouseDown(e) {
      this.$emit("mousedown", e);
    },
    onBlur(e) {
      this.$emit("blur", e);
    },
    onOpen(e) {
      this.$emit("open", e);
    },
    onClose(e) {
      this.$emit("close", e);
    }
  },
  render() {
    const e = v(this), {
      width: n,
      dir: i,
      popupSettings: s
    } = this.$props, {
      popupClass: p,
      className: r,
      animate: u,
      anchor: l,
      anchorAlign: d,
      show: a,
      itemsCount: y,
      appendTo: h,
      collision: c,
      direction: m,
      offset: f,
      popupAlign: w,
      style: g,
      width: o,
      popupStyle: C
    } = s;
    return t(b, {
      ref: "popup",
      style: {
        ...g,
        width: o || n,
        direction: i
      },
      popupClass: r,
      animate: u,
      anchor: l,
      anchorAlign: d,
      show: a,
      onOpen: this.onOpen,
      onClose: this.onClose,
      appendTo: h,
      collision: c,
      direction: m,
      offset: f,
      popupAlign: w
    }, {
      default: () => [t("div", {
        class: p,
        style: C,
        onMousedown: this.onMouseDown,
        onFocusout: this.onBlur
      }, [e])]
    });
  }
});
export {
  O as ListContainer
};
