// @progress/kendo-vue-dropdowns/common/ListFilter.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2025 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
import { defineComponent as c, createVNode as o, ref as p } from "vue";
import { getRef as f, Icon as m, kendoThemeMaps as h, setRef as v } from "@progress/kendo-vue-common";
import { searchIcon as g } from "@progress/kendo-svg-icons";
import { TextBox as b } from "@progress/kendo-vue-inputs";
import { provideLocalizationService as x } from "@progress/kendo-vue-intl";
import { filterLabel as r, messages as k } from "../messages/main.mjs";
const {
  sizeMap: y,
  roundedMap: I
} = h, R = /* @__PURE__ */ c({
  name: "list-filter",
  emits: {
    keydown: (e) => !0,
    change: (e) => !0,
    focus: (e) => !0,
    blur: (e) => !0
  },
  inject: {
    kendoLocalizationService: {
      default: null
    }
  },
  props: {
    value: String,
    rounded: {
      type: String,
      default: "medium",
      validator: function(e) {
        return ["small", "medium", "large", "full"].includes(e);
      }
    },
    fillMode: {
      type: String,
      default: "solid",
      validator: function(e) {
        return ["solid", "flat", "outline"].includes(e);
      }
    },
    size: {
      type: String,
      default: "medium",
      validator: function(e) {
        return ["small", "medium", "large"].includes(e);
      }
    },
    adaptiveMode: {
      type: Boolean,
      default: !1
    },
    ariaControlsId: String,
    ariaActivedescendantId: String,
    tabIndex: {
      type: Number,
      default: 0
    }
  },
  computed: {
    spanClass() {
      const {
        size: e,
        rounded: t,
        fillMode: n
      } = this.$props;
      return {
        "k-searchbox k-input": !0,
        [`k-input-${y[e] || e}`]: e,
        [`k-rounded-${I[t] || t}`]: t,
        [`k-input-${n}`]: n
      };
    }
  },
  setup() {
    return {
      inputRef: p(null)
    };
  },
  methods: {
    onKeyDown(e) {
      this.$emit("keydown", e);
    },
    onChange(e) {
      this.$emit("change", e);
    },
    handleFocus(e) {
      this.$emit("focus", e);
    },
    handleBlur(e) {
      this.$emit("blur", e);
    },
    prefixRender(e) {
      return e(m, {
        ...{
          class: "k-input-icon",
          name: "search",
          icon: g
        }
      });
    }
  },
  mounted() {
    this.input = f(this, "input")._input;
  },
  render() {
    const {
      size: e,
      rounded: t,
      fillMode: n,
      adaptiveMode: a,
      ariaControlsId: s,
      ariaActivedescendantId: l,
      tabIndex: u
    } = this.$props, d = x(this).toLanguageString(r, k[r]), i = function() {
      return o(b, {
        ref: v(this, "input"),
        value: this.$props.value,
        onInput: this.onChange,
        onKeydown: this.onKeyDown,
        onFocus: this.handleFocus,
        onBlur: this.handleBlur,
        size: e,
        fillMode: n,
        rounded: t,
        inputPrefix: this.prefixRender,
        role: "searchbox",
        "aria-label": d,
        "aria-haspopup": "listbox",
        "aria-autocomplete": "list",
        "aria-activedescendant": l,
        "aria-controls": s,
        tabIndex: u
      }, null);
    };
    return a ? i.call(this) : o("div", {
      class: "k-list-filter"
    }, [i.call(this)]);
  }
});
export {
  R as ListFilter
};
