// @progress/kendo-vue-dropdowns/common/ListItem.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2025 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
import { defineComponent as i, createVNode as e, h as o } from "vue";
import { classNames as r, getTemplate as l } from "@progress/kendo-vue-common";
import { getItemValue as n } from "./utils.mjs";
const c = /* @__PURE__ */ i({
  name: "KendoListItem",
  inheritAttrs: !1,
  props: {
    id: String,
    index: Number,
    dataItem: [Object, String, Number],
    textField: String,
    focused: Boolean,
    selected: Boolean,
    render: Object,
    onItemClick: Function
  },
  computed: {
    itemClass() {
      return {
        "k-list-item": !0,
        "k-selected": this.$props.selected,
        "k-focus": this.$props.focused
      };
    }
  },
  methods: {
    handleClick(t) {
      this.$emit("itemClick", this.$props.index, t);
    }
  },
  render() {
    const {
      selected: t
    } = this.$props, s = e("li", {
      id: this.$props.id,
      role: "option",
      "aria-selected": t,
      class: r("k-list-item", {
        "k-selected": t,
        "k-focus": this.$props.focused,
        [this.$attrs.class]: this.$attrs.class
      }),
      onClick: this.handleClick,
      style: this.$attrs.style || {
        position: "unset"
      }
    }, [e("span", {
      class: "k-list-item-text"
    }, [n(this.$props.dataItem, this.$props.textField).toString()])]);
    return l.call(this, {
      h: o,
      template: this.$props.render,
      defaultRendering: s,
      additionalProps: {
        ...this.$props,
        itemClass: this.itemClass
      },
      additionalListeners: {
        click: this.handleClick
      }
    });
  }
});
export {
  c as ListItem
};
