// @progress/kendo-vue-dropdowns/common/List.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2025 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
import { defineComponent as D, createVNode as i, ref as N, inject as B, h as G } from "vue";
import { ListItem as V } from "./ListItem.mjs";
import { ListGroupItem as E } from "./ListGroupItem.mjs";
import { areSame as v, getItemValue as k } from "./utils.mjs";
import { getDefaultSlots as H, setRef as T, getRef as q, getTemplate as J } from "@progress/kendo-vue-common";
import { provideLocalizationService as K } from "@progress/kendo-vue-intl";
import { nodata as y, messages as M } from "../messages/main.mjs";
const te = /* @__PURE__ */ D({
  name: "list",
  emits: {
    listclick: (e) => !0,
    scroll: (e) => !0
  },
  inheritAttrs: !1,
  props: {
    id: String,
    show: Boolean,
    dataItems: Array,
    value: [Object, String, Number, Boolean, Array],
    textField: String,
    valueField: String,
    optionsGuid: String,
    wrapperCssClass: String,
    wrapperStyle: Object,
    listStyle: Object,
    skip: Number,
    focusedIndex: Number,
    highlightSelected: {
      type: Boolean,
      default: !0
    },
    itemRender: [String, Function, Object],
    groupHeaderItemRender: [String, Function, Object],
    noDataRender: [String, Function, Object],
    scroller: Boolean,
    groupField: String
  },
  inject: {
    kendoLocalizationService: {
      default: null
    }
  },
  setup() {
    const e = N(null), r = B("kendoLocalizationService", {});
    return {
      listRef: e,
      kendoLocalizationService: r
    };
  },
  mounted() {
    this.list = q(this, "list");
  },
  methods: {
    handleClick(e, r) {
      this.$emit("listclick", e, r);
    },
    handleScroll(e) {
      this.$emit("scroll", e);
    }
  },
  render() {
    const e = H(this), r = K(this), {
      id: I,
      show: b,
      wrapperCssClass: R,
      wrapperStyle: j,
      listStyle: F,
      listRef: P
    } = this.$props, C = function() {
      const c = this.$props.noDataRender, l = i("div", {
        class: "k-nodata"
      }, [i("div", null, [r.toLanguageString(y, M[y])])]);
      return J.call(this, {
        h: G,
        template: c,
        defaultRendering: l
      });
    }, f = function() {
      const {
        textField: c,
        valueField: l,
        optionsGuid: h,
        skip: $ = 0,
        focusedIndex: x,
        highlightSelected: L,
        value: u,
        groupField: o,
        itemRender: O,
        groupHeaderItemRender: w,
        dataItems: g
      } = this.$props, S = Array.isArray(u);
      let s = 0;
      return g.map(function(t, p) {
        const n = $ + p, A = L && (!S && v(t, u, l) || S && u.findIndex((z) => v(z, t, l)) !== -1);
        let a, d, m;
        return p > 0 && o !== void 0 && (d = k(t, o), m = k(g[p - 1], o), d && m && d !== m !== void 0 && (a = d)), a !== void 0 && Object.keys(t).length === 1 && (s += 1), a !== void 0 && t[o] !== void 0 && Object.keys(t).length === 1 ? i(E, {
          id: `option-${h}-${n}`,
          key: n + "-group-item",
          group: a,
          render: w
        }, null) : i(V, {
          id: `option-${h}-${n}`,
          dataItem: t,
          selected: A,
          focused: x === n - s,
          index: n - s,
          key: n - s,
          onItemClick: this.handleClick,
          textField: c,
          render: O
        }, null);
      }, this);
    }.call(this);
    return f.length ? i("div", {
      class: R,
      style: j,
      unselectable: "on",
      onScroll: this.handleScroll
    }, [i("ul", {
      id: I,
      role: "listbox",
      "aria-live": "polite",
      "aria-hidden": b ? void 0 : !0,
      class: "k-list-ul",
      ref: T(this, "list"),
      style: F
    }, [f]), e]) : C.call(this);
  }
});
export {
  te as List
};
