// @progress/kendo-vue-dropdowns/common/SearchBar.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2025 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
import { defineComponent as r, createVNode as o, mergeProps as d, ref as p } from "vue";
import { setRef as u, getRef as l, canUseDOM as h } from "@progress/kendo-vue-common";
const b = /* @__PURE__ */ r({
  name: "search-bar",
  emits: {
    change: (e) => !0,
    keydown: (e) => !0,
    focus: (e) => !0,
    blur: (e) => !0
  },
  props: {
    value: String,
    id: String,
    placeholder: String,
    tabIndex: Number,
    size: {
      type: Number,
      default: void 0
    },
    suggestedText: String,
    focused: Boolean,
    disabled: Boolean,
    readOnly: Boolean,
    expanded: Boolean,
    owns: String,
    name: String,
    activedescendant: String,
    ariaDescribedBy: String,
    clearButton: Boolean,
    accessKey: String,
    ariaLabelledBy: String,
    ariaLabel: {
      type: String,
      default: void 0
    },
    onClearbuttonclick: Function,
    inputAttributes: Object
  },
  data() {
    return {
      prevValue: void 0,
      prevSuggestedText: void 0
    };
  },
  watch: {
    suggestedText: function(e, t) {
      this.prevSuggestedText = t;
    },
    value: function(e, t) {
      this.prevValue = t;
    }
  },
  updated() {
    const {
      value: e,
      suggestedText: t,
      focused: n
    } = this.$props;
    this.input || (this.input = this.$refs.input);
    const i = this.input, s = this.$data.prevValue !== e || t !== this.prevSuggestedText, a = s && this.$data.prevValue && this.$data.prevValue.startsWith(e) && !(this.$data.prevSuggestedText && t && this.$data.prevSuggestedText.endsWith(t));
    n && i && h && document.activeElement !== i && i.focus(), t && s && !a && i && i.setSelectionRange(e.length - t.length, e.length);
  },
  setup() {
    return {
      inputRef: p(null)
    };
  },
  mounted() {
    this.input = l(this, "input");
  },
  methods: {
    onChange(e) {
      this.$emit("change", e);
    },
    onBlur(e) {
      this.$emit("blur", e);
    },
    onFocus(e) {
      this.$emit("focus", e);
    },
    onKeyDown(e) {
      this.$emit("keydown", e);
    },
    clearButtonClick(e) {
      this.$emit("clearbuttonclick", e);
    }
  },
  render() {
    return o("input", d({
      autocomplete: "off",
      id: this.$props.id,
      type: "text",
      key: "searchbar",
      placeholder: this.$props.placeholder,
      class: "k-input-inner",
      tabindex: this.$props.tabIndex,
      accesskey: this.$props.accessKey,
      role: "combobox",
      name: this.$props.name,
      value: this.$props.value,
      size: this.$props.size ? this.$props.size : 20,
      onInput: this.onChange,
      ref: u(this, "input"),
      onKeydown: this.onKeyDown,
      onFocusin: this.onFocus,
      onBlur: this.onBlur,
      "aria-disabled": this.$props.disabled || void 0,
      disabled: this.$props.disabled || void 0,
      readonly: this.$props.readOnly || void 0,
      "aria-autocomplete": this.$props.suggestedText ? "both" : "list",
      "aria-haspopup": "listbox",
      "aria-expanded": this.$props.expanded || !1,
      "aria-controls": this.$props.owns,
      "aria-activedescendant": this.$props.expanded ? this.$props.activedescendant : void 0,
      "aria-describedby": this.$props.ariaDescribedBy,
      "aria-labelledby": this.$props.ariaLabelledBy,
      "aria-label": this.$props.ariaLabel
    }, this.$props.inputAttributes), null);
  }
});
export {
  b as SearchBar
};
