// @progress/kendo-vue-grid/columnMenu/GridColumnMenuFilterOperators.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2025 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
import { defineComponent as o, createVNode as p } from "vue";
import { DropDownList as a } from "@progress/kendo-vue-dropdowns";
import { cellOperatorChange as n } from "../filterCommon.mjs";
const h = /* @__PURE__ */ o({
  props: {
    filterType: String,
    operator: String,
    operators: Array
  },
  emits: {
    change: null
  },
  methods: {
    triggerChange(r) {
      this.$emit("change", r);
    },
    operatorChange(r, e) {
      const t = n(r.value.operator, e, this.$props.value);
      this.triggerChange(t);
    }
  },
  render() {
    const r = this.$props.operators.find((e) => e.operator === this.$props.operator) || null;
    return this.$props.filterType !== "boolean" ? p(a, {
      onChange: this.operatorChange,
      value: r,
      "data-items": this.$props.operators,
      textField: "text"
    }, null) : null;
  }
});
export {
  h as ColumnMenuFilterOperators
};
