// @progress/kendo-vue-layout/appbar/AppBarSpacer.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2025 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
import { defineComponent as s, createVNode as t } from "vue";
import { focusFirstFocusableChild as r, getDefaultSlots as a } from "@progress/kendo-vue-common";
const d = /* @__PURE__ */ s({
  name: "KendoAppBarSpacer",
  props: {
    width: [String, Number]
  },
  computed: {
    spacerClasses() {
      return {
        "k-appbar-spacer": !0,
        "k-appbar-spacer-sized": this.width !== void 0
      };
    },
    spacerStyles() {
      return {
        flexBasis: this.width !== void 0 ? this.width : void 0
      };
    }
  },
  render() {
    const e = a(this);
    return t("span", {
      class: this.spacerClasses,
      style: this.spacerStyles
    }, [e]);
  },
  methods: {
    focus() {
      this.$el && r(this.$el);
    }
  }
});
export {
  d as AppBarSpacer
};
