// @progress/kendo-vue-animation/Expand.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2025 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
import { defineComponent as o, createVNode as i, isVNode as n } from "vue";
import { Animation as r } from "./Animation.mjs";
import { getDefaultSlots as a } from "@progress/kendo-vue-common";
function p(t) {
  return typeof t == "function" || Object.prototype.toString.call(t) === "[object Object]" && !n(t);
}
const f = /* @__PURE__ */ o({
  props: {
    appear: {
      type: Boolean,
      default: !1
    },
    enter: {
      type: Boolean,
      default: !0
    },
    exit: {
      type: Boolean,
      default: !0
    },
    transitionEnterDuration: {
      type: Number,
      default: 300
    },
    transitionExitDuration: {
      type: Number,
      default: 300
    },
    childFactory: Object,
    className: String,
    direction: {
      type: String,
      default: "vertical"
    },
    componentChildStyle: Object,
    tag: String,
    id: String
  },
  render() {
    const t = a(this), {
      direction: s,
      ...l
    } = this.$props, e = `expand-${this.$props.direction}`;
    return i(r, {
      transitionName: e,
      appear: this.$props.appear,
      enter: this.$props.enter,
      exit: this.$props.exit,
      componentChildStyle: this.$props.componentChildStyle,
      transitionEnterDuration: this.$props.transitionEnterDuration,
      transitionExitDuration: this.$props.transitionExitDuration
    }, p(t) ? t : {
      default: () => [t]
    });
  }
});
export {
  f as Expand
};
