// @progress/kendo-vue-animation/Fade.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2025 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
import { defineComponent as e, createVNode as o, isVNode as n } from "vue";
import { Animation as i } from "./Animation.mjs";
import { getDefaultSlots as r } from "@progress/kendo-vue-common";
function a(t) {
  return typeof t == "function" || Object.prototype.toString.call(t) === "[object Object]" && !n(t);
}
const u = /* @__PURE__ */ e({
  props: {
    appear: {
      type: Boolean,
      default: !1
    },
    enter: {
      type: Boolean,
      default: !0
    },
    exit: {
      type: Boolean,
      default: !1
    },
    transitionEnterDuration: {
      type: Number,
      default: 500
    },
    transitionExitDuration: {
      type: Number,
      default: 500
    },
    componentChildStyle: Object,
    childFactory: Object,
    className: String,
    tag: String,
    id: String
  },
  render() {
    const t = r(this);
    return o(i, {
      transitionName: "fade",
      appear: this.$props.appear,
      enter: this.$props.enter,
      exit: this.$props.exit,
      componentChildStyle: this.$props.componentChildStyle,
      transitionEnterDuration: this.$props.transitionEnterDuration,
      transitionExitDuration: this.$props.transitionExitDuration
    }, a(t) ? t : {
      default: () => [t]
    });
  }
});
export {
  u as Fade
};
