// @progress/kendo-vue-animation/Push.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2025 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
import { defineComponent as i, createVNode as o, isVNode as n } from "vue";
import { Animation as r } from "./Animation.mjs";
import { getDefaultSlots as a } from "@progress/kendo-vue-common";
function p(t) {
  return typeof t == "function" || Object.prototype.toString.call(t) === "[object Object]" && !n(t);
}
const s = {
  position: "absolute",
  top: "0",
  left: "0"
}, d = /* @__PURE__ */ i({
  props: {
    appear: {
      type: Boolean,
      default: !1
    },
    enter: {
      type: Boolean,
      default: !0
    },
    exit: {
      type: Boolean,
      default: !0
    },
    transitionEnterDuration: {
      type: Number,
      default: 300
    },
    transitionExitDuration: {
      type: Number,
      default: 300
    },
    childFactory: Object,
    componentChildStyle: Object,
    className: String,
    direction: {
      type: String,
      default: "right"
    },
    tag: String,
    id: String
  },
  render() {
    const t = a(this), e = `push-${this.$props.direction}`;
    return o(r, {
      transitionName: e,
      appear: this.$props.appear,
      enter: this.$props.enter,
      exit: this.$props.exit,
      transitionEnterDuration: this.$props.transitionEnterDuration,
      transitionExitDuration: this.$props.transitionExitDuration,
      componentChildStyle: this.$props.componentChildStyle,
      animationExitingStyle: this.$props.stackChildren ? s : void 0
    }, p(t) ? t : {
      default: () => [t]
    });
  }
});
export {
  d as Push
};
