// @progress/kendo-vue-progressbars/chunkprogressbar/ChunkProgressBar.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2025 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
import { defineComponent as z, createVNode as n } from "vue";
import { getTabIndex as B, classNames as h, getDir as D, validatePackage as I } from "@progress/kendo-vue-common";
import { calculateRatio as j } from "../common/utils.mjs";
import { packageMetadata as w } from "../package-metadata.mjs";
const E = /* @__PURE__ */ z({
  name: "KendoChunkProgressBar",
  props: {
    chunkCount: {
      type: Number,
      default: 5
    },
    ariaLabel: String,
    disabled: Boolean,
    reverse: {
      type: Boolean,
      default: !1
    },
    max: {
      type: Number,
      default: 100
    },
    min: {
      type: Number,
      default: 0
    },
    value: {
      type: Number,
      default: 0
    },
    tabIndex: Number,
    emptyStyle: Object,
    emptyClassName: String,
    progressStyle: Object,
    progressClassName: String,
    orientation: {
      type: String,
      default: "horizontal",
      validator: function(s) {
        return ["horizontal", "vertical"].includes(s);
      }
    }
  },
  created() {
    I(w);
  },
  data() {
    return {
      currentDir: void 0
    };
  },
  mounted() {
    this.currentDir = D(this.$el, this.$props.dir);
  },
  render() {
    const {
      chunkCount: s,
      disabled: a,
      orientation: k,
      min: l,
      max: i,
      reverse: b,
      tabIndex: f,
      emptyStyle: g,
      emptyClassName: y,
      progressStyle: v,
      progressClassName: C
    } = this.$props, u = this.$props.value, d = this.$props.value === null, t = k === "vertical", S = function(r) {
      let c = [];
      const p = 100 / r + "%", N = j(l, i, u), x = Math.floor(N * r), m = Array(r).fill(!1);
      for (let e = 0; e < x; e++)
        m[e] = !0;
      for (let e = 0; e < r; ++e) {
        const o = m[e], $ = o ? C : y, P = {
          ...{
            width: t ? void 0 : p,
            height: t ? p : void 0
          },
          ...o ? v : g
        };
        c.push(n("li", {
          key: e,
          class: h("k-progressbar-chunk", {
            "k-first": e === 0,
            "k-last": e === r - 1,
            "k-selected": o
          }, $),
          style: P
        }, null));
      }
      return c;
    };
    return n("div", {
      class: h("k-progressbar", "k-chunk-progressbar", {
        "k-progressbar-horizontal": !t,
        "k-progressbar-vertical": t,
        "k-progressbar-reverse": b,
        "k-progressbar-indeterminate": d,
        "k-disabled": a
      }),
      dir: this.currentDir,
      tabindex: B(f, a),
      role: "progressbar",
      "aria-label": this.$props.ariaLabel,
      "aria-valuemin": l,
      "aria-valuemax": i,
      "aria-valuenow": d ? void 0 : u,
      "aria-disabled": a
    }, [n("ul", {
      class: "k-progressbar-chunks k-reset"
    }, [S.call(this, s)])]);
  },
  methods: {
    focusElement() {
      this.$el && this.$el.focus();
    }
  }
});
export {
  E as ChunkProgressBar
};
