// @progress/kendo-vue-intl/index.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2025 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
import { provideIntlService as e, provideLocalizationService as i } from "./intlUtils.mjs";
import { IntlProvider as a } from "./Intl/IntlProvider.mjs";
import { IntlService as l } from "./Intl/IntlService.mjs";
import { loadMessages as c } from "./Localization/loadMessages.mjs";
import { LocalizationProvider as m } from "./Localization/LocalizationProvider.mjs";
import { LocalizationService as d } from "./Localization/LocalizationService.mjs";
import { load as S } from "@progress/kendo-intl";
export {
  a as IntlProvider,
  l as IntlService,
  m as LocalizationProvider,
  d as LocalizationService,
  S as load,
  c as loadMessages,
  e as provideIntlService,
  i as provideLocalizationService
};
